{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CART-Microbiome: exploratory data-analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy.stats import mannwhitneyu, wilcoxon\n",
    "from scipy.spatial import distance\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from scipy.interpolate import make_interp_spline\n",
    "\n",
    "from skbio.diversity.alpha import shannon\n",
    "from skbio.stats.distance import permanova\n",
    "from skbio import DistanceMatrix\n",
    "from skbio.stats.ordination import pcoa\n",
    "\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from matplotlib.legend_handler import HandlerBase\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Defining directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_input = \"../../CART_Microbiome_input/\"\n",
    "path_output = \"../../CART_Microbiome_output/\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading and processing the input tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_CART_table_stool =      pd.read_excel(path_input + 'CART_table_stool.xlsx')\n",
    "df_CART_table_taxa =       pd.read_excel(path_input + 'CART_table_taxa.xlsx')\n",
    "df_CART_table_pathways =   pd.read_excel(path_input + 'CART_table_pathways.xlsx')\n",
    "df_CART_table_genes =      pd.read_excel(path_input + 'CART_table_genes.xlsx')\n",
    "df_CART_table_patients =   pd.read_excel(path_input + 'CART_table_patients.xlsx')\n",
    "df_CART_table_response =   pd.read_excel(path_input + 'CART_table_response.xlsx')\n",
    "df_CART_table_blood =      pd.read_excel(path_input + 'CART_table_blood.xlsx')\n",
    "CR_coefficients_df =       pd.read_excel(path_input + 'CR180_noHighRiskABX_elasticnet_train_coefficients.xlsx')\n",
    "df_kdt_numbers =           pd.read_excel(path_input + 'CART_postkdt_reads.xlsx')\n",
    "\n",
    "\n",
    "df_CART_table_stool.index = df_CART_table_stool.stool_sample_id\n",
    "df_CART_table_taxa.index = df_CART_table_taxa.stool_sample_id\n",
    "df_CART_table_pathways.index = df_CART_table_pathways.stool_sample_id\n",
    "df_CART_table_genes.index = df_CART_table_genes.stool_sample_id\n",
    "\n",
    "df_CART_table_taxa = df_CART_table_taxa.drop([\"stool_sample_id\"], axis = 1)\n",
    "df_CART_table_pathways = df_CART_table_pathways.drop([\"stool_sample_id\",\"UNMAPPED\", \"UNINTEGRATED\"], axis = 1)\n",
    "df_CART_table_genes = df_CART_table_genes.drop([\"stool_sample_id\",\"UNMAPPED\", \"UNGROUPED\"], axis = 1)\n",
    "\n",
    "\n",
    "# Identifying patients who have a mixture of both HR-Abx and none-|LR-Abx basal samples:\n",
    "\n",
    "df_CART_table_stool_post_basal_removed = df_CART_table_stool.loc[(df_CART_table_stool[\"sampleDay_sample\"] < 1) &  (df_CART_table_stool[\"sampleDay_sample\"] > -22),:]\n",
    "patient_mixed_basal = (df_CART_table_stool_post_basal_removed.groupby(\"subjectID\")[\"Basal_no_HighRisk_sample\"].apply(lambda x:x.value_counts()).to_frame().groupby(level=\"subjectID\").size()==2)\n",
    "patient_mixed_basal_true = list(patient_mixed_basal[patient_mixed_basal ==True].index)\n",
    "\n",
    "\n",
    "# Identifying samples that have < 5e5 post-KneadData reads:\n",
    "\n",
    "df_kdt_numbers = df_kdt_numbers[~df_kdt_numbers.FileName.str.contains(\"paired_2\")] \n",
    "df_kdt_numbers['FileName'] = df_kdt_numbers['FileName'].str.split(\"_combined_R1_001_kneaddata\").str.get(0)#\n",
    "df_kdt_numbers['FileName'] = df_kdt_numbers['FileName'].str.split(\"-LR-5\").str.get(0)\n",
    "\n",
    "df_kdt_numbers['FileName'] = df_kdt_numbers['FileName'].str.split(\"-LR-5\").str.get(0)\n",
    "df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('CART', na=False)==True,'FileName'] = df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('CART', na=False)==True,'FileName'].str.split(\"_S\").str.get(0)\n",
    "df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('S00AF', na=False)==True,'FileName'] = df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('S00AF', na=False)==True,'FileName'].str.split(\"_S\").str.get(0)\n",
    "df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('MDA', na=False)==True,'FileName'] = df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('MDA', na=False)==True,'FileName'].str.split(\"_S\").str.get(0)\n",
    "df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('19696', na=False)==True,'FileName'] = df_kdt_numbers.loc[df_kdt_numbers['FileName'].str.startswith('19696', na=False)==True,'FileName'].str.split(\"_S\").str.get(0)\n",
    "\n",
    "df_kdt_numbers.index = df_kdt_numbers.FileName\n",
    "common_names_kdt = set(df_kdt_numbers.index).intersection(set(df_CART_table_stool.stool_sample_id))\n",
    "\n",
    "df_kdt_numbers_filtered = df_kdt_numbers.loc[common_names_kdt,:]\n",
    "\n",
    "samples_below_thresholds = list(df_kdt_numbers_filtered.loc[df_kdt_numbers_filtered[' Number of reads kdt'] < 5e5].FileName)\n",
    "\n",
    "# Patients exclusion:\n",
    "\n",
    "df_CART_table_patients_filtered = df_CART_table_patients.loc[(df_CART_table_patients.subjectID).isin(df_CART_table_stool.subjectID) & # Excluding patinets due to lack of stool samples\n",
    "                                                             ~(df_CART_table_patients.subjectID).isin(patient_mixed_basal_true) & # Excluding patinets due to mixed pre-infusing Abx. risks\n",
    "                                                             ~df_CART_table_patients.Progression_early.isnull(),:] # Excluding patients due to an incomplete follow-up.\n",
    "\n",
    "df_CART_table_patients_filtered.index = df_CART_table_patients_filtered.subjectID\n",
    "\n",
    "\n",
    "# Identifying samples that do not belong to any included patient:\n",
    "\n",
    "creteria_donor_included = df_CART_table_stool[\"subjectID\"].isin(df_CART_table_patients_filtered.subjectID)\n",
    "\n",
    "# Defining the criteria for samples inclusion:\n",
    "\n",
    "creteria_stool_samples = creteria_donor_included & ~df_CART_table_stool.stool_sample_id.isin(samples_below_thresholds)\n",
    "\n",
    "\n",
    "# Defining the final processed and filtered tables:\n",
    "\n",
    "df_CART_table_stool_filtered = df_CART_table_stool.loc[creteria_stool_samples,:]\n",
    "df_CART_table_taxa_filtered =  df_CART_table_taxa.loc[creteria_stool_samples,:]\n",
    "df_CART_table_pathways_filtered = df_CART_table_pathways.loc[creteria_stool_samples,:]\n",
    "df_CART_table_genes_filtered = df_CART_table_genes.loc[creteria_stool_samples,:]\n",
    "\n",
    "df_CART_table_response_filtered = df_CART_table_response.loc[df_CART_table_response[\"subjectID\"].isin(df_CART_table_patients_filtered.subjectID),:]\n",
    "df_CART_table_blood_filtered = df_CART_table_blood.loc[df_CART_table_blood[\"subjectID\"].isin(df_CART_table_patients_filtered.subjectID),:]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsetting the samples based on the relative sampling day and the pre-infusion antibiotics-risk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"all\":\n",
    "\n",
    "df_CART_table_stool_all =    df_CART_table_stool_filtered.copy()\n",
    "df_CART_table_taxa_all =     df_CART_table_taxa_filtered.copy()\n",
    "df_CART_table_pathways_all = df_CART_table_pathways_filtered.copy()\n",
    "df_CART_table_genes_all =    df_CART_table_genes_filtered.copy()\n",
    "\n",
    "\n",
    "# \"basal_all\":\n",
    "\n",
    "creteria_basal = (df_CART_table_stool_filtered[\"sampleDay_sample\"] < 1) &  (df_CART_table_stool_filtered[\"sampleDay_sample\"] > -22)\n",
    "\n",
    "df_CART_table_stool_basal_all = df_CART_table_stool_filtered.loc[creteria_basal,:]\n",
    "df_CART_table_taxa_basal_all = df_CART_table_taxa_filtered.loc[creteria_basal,:]\n",
    "df_CART_table_pathways_basal_all = df_CART_table_pathways_filtered.loc[creteria_basal,:]\n",
    "df_CART_table_genes_basal_all = df_CART_table_genes_filtered.loc[creteria_basal,:]\n",
    "\n",
    "\n",
    "# \"basal_noHighRiskAbx\":\n",
    "\n",
    "creteria_NoHighRisk = df_CART_table_stool_basal_all[\"Basal_no_HighRisk_sample\"] == True\n",
    "\n",
    "df_CART_table_stool_basal_NoHighRisk =    df_CART_table_stool_basal_all.loc[creteria_NoHighRisk,:]\n",
    "df_CART_table_taxa_basal_NoHighRisk =     df_CART_table_taxa_basal_all.loc[creteria_NoHighRisk,:]\n",
    "df_CART_table_pathways_basal_NoHighRisk = df_CART_table_pathways_basal_all.loc[creteria_NoHighRisk,:]\n",
    "df_CART_table_genes_basal_NoHighRisk =    df_CART_table_genes_basal_all.loc[creteria_NoHighRisk,:]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Deriving the mean-per patient basal microbiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def derive_mean_microbiome(microbiome_df, samples_df):\n",
    "    \n",
    "    _microbiome_df = microbiome_df.copy()\n",
    "    _microbiome_df = _microbiome_df.div(_microbiome_df.sum(axis=1), axis=0)\n",
    "    \n",
    "    microbiome_df_mean = samples_df.loc[:, ['subjectID']].join(_microbiome_df).groupby(\"subjectID\")[_microbiome_df.columns].mean()\n",
    "    \n",
    "    samples_df_mean = samples_df.groupby(samples_df[\"subjectID\"]).apply(lambda _df : _df.iloc[0,:]) \n",
    "    samples_df_mean.index = samples_df_mean.subjectID\n",
    "\n",
    "    return(microbiome_df_mean, samples_df_mean)\n",
    "\n",
    "\n",
    "# basal_all_mean:\n",
    "\n",
    "df_CART_table_stool_basal_all_mean =    derive_mean_microbiome(df_CART_table_taxa_basal_all, df_CART_table_stool_basal_all)[1]\n",
    "df_CART_table_taxa_basal_all_mean =     derive_mean_microbiome(df_CART_table_taxa_basal_all, df_CART_table_stool_basal_all)[0]\n",
    "df_CART_table_pathways_basal_all_mean = derive_mean_microbiome(df_CART_table_pathways_basal_all, df_CART_table_stool_basal_all)[0]\n",
    "df_CART_table_genes_basal_all_mean =    derive_mean_microbiome(df_CART_table_genes_basal_all, df_CART_table_stool_basal_all)[0]\n",
    "\n",
    "\n",
    "# basal_NoHighRisk_mean:\n",
    "\n",
    "df_CART_table_stool_basal_NoHighRisk_mean = df_CART_table_stool_basal_all_mean.loc[df_CART_table_stool_basal_all_mean[\"Basal_no_HighRisk_sample\"]]\n",
    "df_CART_table_taxa_basal_NoHighRisk_mean = df_CART_table_taxa_basal_all_mean.loc[df_CART_table_stool_basal_all_mean[\"Basal_no_HighRisk_sample\"]]\n",
    "df_CART_table_pathways_basal_NoHighRisk_mean = df_CART_table_pathways_basal_all_mean.loc[df_CART_table_stool_basal_all_mean[\"Basal_no_HighRisk_sample\"]]\n",
    "df_CART_table_genes_basal_NoHighRisk_mean = df_CART_table_genes_basal_all_mean.loc[df_CART_table_stool_basal_all_mean[\"Basal_no_HighRisk_sample\"]]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Defining the function \"abundance_filtration\" and the abundance thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa_abundance_threshold = 1e-3\n",
    "pathways_abundance_threshold = 1e-4\n",
    "genes_abundance_threshold = 1e-4\n",
    "\n",
    "def abundance_filtration(df_X, abundnace_threshold):\n",
    "    \n",
    "    _df_X = df_X.copy()\n",
    "    _df_X = _df_X.div(_df_X.sum(axis=1), axis=0)\n",
    "\n",
    "    _df_X[_df_X <= abundnace_threshold] = abundnace_threshold\n",
    "    _df_X = _df_X.loc[:,_df_X.max() > abundnace_threshold]\n",
    "    \n",
    "    return _df_X.div(_df_X.sum(axis=1), axis=0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculating Shannon indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shannon for all samples: \n",
    "\n",
    "df_CART_table_taxa_all_norm = abundance_filtration(df_CART_table_taxa_all, taxa_abundance_threshold)\n",
    "\n",
    "df_Shannon = pd.DataFrame(index=df_CART_table_taxa_all_norm.index, columns=[\"Shannon\"])\n",
    "                          \n",
    "for current_sample in df_Shannon.index:\n",
    "    df_Shannon.loc[current_sample, \"Shannon\"] = shannon(df_CART_table_taxa_all_norm.loc[current_sample,:])\n",
    "    \n",
    "\n",
    "# Shannon for the mean basal microbiome:\n",
    "\n",
    "df_Shannon_basal_mean_all = pd.DataFrame(index=df_CART_table_taxa_basal_all_mean.index, columns=[\"Shannon\"])\n",
    "\n",
    "for current_subject in df_Shannon_basal_mean_all.index:\n",
    "    df_Shannon_basal_mean_all.loc[current_subject, \"Shannon\"] = shannon(df_CART_table_taxa_basal_all_mean.loc[current_subject,:])\n",
    "\n",
    "\n",
    "# Shannon for the NoHighRisk mean basal microbiome: \n",
    "\n",
    "df_Shannon_basal_mean_NoHighRisk = pd.DataFrame(index=df_CART_table_taxa_basal_NoHighRisk_mean.index, columns=[\"Shannon\"])\n",
    "\n",
    "for current_subject in df_Shannon_basal_mean_NoHighRisk.index:\n",
    "    df_Shannon_basal_mean_NoHighRisk.loc[current_subject, \"Shannon\"] = shannon(df_CART_table_taxa_basal_NoHighRisk_mean.loc[current_subject,:])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Defining stratifications and graphical-layouts for metadata parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_colvecs = pd.DataFrame(index=df_CART_table_stool_all.index)\n",
    "df_colvecs[\"subjectID\"] = df_CART_table_stool_all[\"subjectID\"] \n",
    "\n",
    "# Patients parameters:\n",
    "\n",
    "for current_patient in df_colvecs.subjectID:\n",
    "\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"Country\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Country\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"Center\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Center\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"Gender\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Gender\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"Age\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Age\"].values\n",
    "    \n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] == current_patient, \"CR180\"] = df_CART_table_response_filtered.loc[(df_CART_table_response_filtered[\"subjectID\"] == current_patient) & (df_CART_table_response_filtered[\"Day\"] == 180) ,\"Response\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"Progression_early\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Progression_early\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"CRS_grade\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CRS_grade\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"ICANS_grade\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"ICANS_grade\"].values\n",
    "  \n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] == current_patient, \"LDH_Depletion\"] = df_CART_table_blood_filtered.loc[(df_CART_table_blood_filtered[\"subjectID\"] == current_patient) & (df_CART_table_blood_filtered[\"Day\"] == \"Depletion\") ,\"LDH\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] == current_patient, \"CRP_depletion\"] = df_CART_table_blood_filtered.loc[(df_CART_table_blood_filtered[\"subjectID\"] == current_patient) & (df_CART_table_blood_filtered[\"Day\"] == \"Depletion\") ,\"CRP\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] == current_patient, \"IL6_depletion\"] = df_CART_table_blood_filtered.loc[(df_CART_table_blood_filtered[\"subjectID\"] == current_patient) & (df_CART_table_blood_filtered[\"Day\"] == \"Depletion\") ,\"IL6\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"CD3_mul_Aph\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CD3_mul_Aph\"].values \n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"CD4_mul_Aph\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CD4_mul_Aph\"].values \n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"CD8_mul_Aph\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CD8_mul_Aph\"].values\n",
    "\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] == current_patient, \"Product\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Product\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"ECOG_grade\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"ECOG\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"BulkyDisease_10cm\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"BulkyDisease_10cm\"].values\n",
    "    df_colvecs.loc[df_colvecs[\"subjectID\"] ==current_patient, \"Prior_Tx\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"Prior_Tx\"].values\n",
    "    \n",
    "# Basal_no_HighRisk_sample:\n",
    "df_colvecs[\"Basal_no_HighRisk_sample\"] = df_CART_table_stool_all[\"Basal_no_HighRisk_sample\"]\n",
    "df_colvecs.loc[df_colvecs[\"Basal_no_HighRisk_sample\"] == False, \"Basal_no_HighRisk_sample_binarized\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"Basal_no_HighRisk_sample\"] == True, \"Basal_no_HighRisk_sample_binarized\"] = 1\n",
    "\n",
    "# AbxRisk_On_Sampling:\n",
    "df_colvecs[\"AbxRisk_On_Sampling\"] = df_CART_table_stool_all[\"AbxRisk_On_Sampling\"]\n",
    "\n",
    "# Shannon:\n",
    "df_colvecs[\"Shannon\"] = df_Shannon[\"Shannon\"]\n",
    "\n",
    "# Progression_early_binarized:\n",
    "df_colvecs.loc[df_colvecs[\"Progression_early\"] == \"no\", \"Progression_early_binarized\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"Progression_early\"] == \"yes\", \"Progression_early_binarized\"] = 1\n",
    "\n",
    "# Age_binarized:\n",
    "df_colvecs.loc[df_colvecs[\"Age\"] < 65, \"Age_binarized\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"Age\"] >= 65, \"Age_binarized\"] = 1\n",
    "\n",
    "# LDH_Depletion_binarized\n",
    "LDH_threshold = 280\n",
    "df_colvecs.loc[df_colvecs[\"LDH_Depletion\"] < LDH_threshold, \"LDH_Depletion_binarized\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"LDH_Depletion\"] >= LDH_threshold, \"LDH_Depletion_binarized\"] = 1\n",
    "\n",
    "# Response_d180_grade\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"] == 'dead', \"Response_d180_grade\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"] == 'PD', \"Response_d180_grade\"] = 1\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"] == 'PR', \"Response_d180_grade\"] = 2\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"] == 'CR', \"Response_d180_grade\"] = 3\n",
    "\n",
    "# Response_binarized\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"].isin([\"PR\", \"PD\", \"dead\"]), \"CR180_binarized\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"] == \"CR\", \"CR180_binarized\"] = 1\n",
    "\n",
    "# Survival\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"] == \"dead\", \"Survival_d180\"] = 0\n",
    "df_colvecs.loc[df_colvecs[\"CR180\"].isin([\"CR\", \"PR\", \"PD\"]), \"Survival_d180\"] = 1\n",
    "\n",
    "# CRS_01 binarized\n",
    "df_colvecs.loc[df_colvecs.CRS_grade < 2 , \"CRS_01\"] = 1\n",
    "df_colvecs.loc[df_colvecs.CRS_grade > 1 , \"CRS_01\"] = 0\n",
    "\n",
    "# CRS 0, 1, 2, 3+4 categorized\n",
    "df_colvecs.loc[df_colvecs.CRS_grade == 0 , \"CRS_012_34\"] = 0\n",
    "df_colvecs.loc[df_colvecs.CRS_grade == 1 , \"CRS_012_34\"] = 1\n",
    "df_colvecs.loc[df_colvecs.CRS_grade == 2 , \"CRS_012_34\"] = 2\n",
    "df_colvecs.loc[df_colvecs.CRS_grade == 3 , \"CRS_012_34\"] = 34\n",
    "df_colvecs.loc[df_colvecs.CRS_grade == 4 , \"CRS_012_34\"] = 34\n",
    "\n",
    "# ICANS_012_34:\n",
    "df_colvecs.loc[df_colvecs.ICANS_grade == 0 , \"ICANS_012_34\"] = 0\n",
    "df_colvecs.loc[df_colvecs.ICANS_grade == 1 , \"ICANS_012_34\"] = 1\n",
    "df_colvecs.loc[df_colvecs.ICANS_grade == 2 , \"ICANS_012_34\"] = 2\n",
    "df_colvecs.loc[df_colvecs.ICANS_grade == 3 , \"ICANS_012_34\"] = 34\n",
    "df_colvecs.loc[df_colvecs.ICANS_grade == 4 , \"ICANS_012_34\"] = 34\n",
    "\n",
    "# ECOG_01, ECOG_01_colvec:\n",
    "df_colvecs[\"ECOG_01_colvec\"] = \"gray\"\n",
    "df_colvecs.loc[df_colvecs.ECOG_grade < 2 , \"ECOG_01\"] = 1\n",
    "df_colvecs.loc[df_colvecs.ECOG_grade < 2 , \"ECOG_01_colvec\"] = \"green\"\n",
    "df_colvecs.loc[df_colvecs.ECOG_grade > 1 , \"ECOG_01\"] = 0\n",
    "df_colvecs.loc[df_colvecs.ECOG_grade > 1 , \"ECOG_01_colvec\"] = \"red\"\n",
    "colvec_label_ECOG_01   = ['ECOG 0-1','ECOG 2-3']\n",
    "colvec_color_ECOG_01  = [\"green\", \"red\"]\n",
    "colvec_marker_ECOG_01    = [\"o\",\"o\",\"o\"]\n",
    "\n",
    "# country_colvec, Country_yesno:\n",
    "df_colvecs[\"country_colvec\"] = \"gray\"\n",
    "df_colvecs.loc[df_colvecs.Country == \"Germany\",\"country_colvec\"] = \"green\"\n",
    "df_colvecs.loc[df_colvecs.Country == \"US\",\"country_colvec\"] = \"red\"\n",
    "df_colvecs.loc[df_colvecs.Country == \"Germany\",\"Country_yesno\"] = \"yes\"\n",
    "df_colvecs.loc[df_colvecs.Country == \"US\",\"Country_yesno\"] = \"no\"\n",
    "colvec_label_country   = ['Germany','US']\n",
    "colvec_color_country  = [\"green\", \"red\"]\n",
    "colvec_marker_country    = [\"o\",\"o\"]\n",
    "\n",
    "# AbxRisk_On_Sampling_colvec:\n",
    "df_colvecs[\"AbxRisk_On_Sampling_colvec\"] = \"gray\"\n",
    "df_colvecs.loc[df_colvecs.AbxRisk_On_Sampling == \"HR\",\"AbxRisk_On_Sampling_colvec\"] = \"orange\"\n",
    "df_colvecs.loc[df_colvecs.AbxRisk_On_Sampling == \"LR\",\"AbxRisk_On_Sampling_colvec\"] = \"lightgreen\"\n",
    "df_colvecs.loc[df_colvecs.AbxRisk_On_Sampling == \"None\",\"AbxRisk_On_Sampling_colvec\"] = \"steelblue\"\n",
    "colvec_label_AbxRisk_On_Sampling   = ['HR','LR', 'None']\n",
    "colvec_color_AbxRisk_On_Sampling  = [\"orange\", \"lightgreen\", \"steelblue\"]\n",
    "colvec_marker_AbxRisk_On_Sampling    = [\"o\",\"o\",\"o\"]\n",
    "\n",
    "# day_colvec:\n",
    "df_colvecs[\"day_colvec\"] = df_CART_table_stool_all.sampleDay_sample\n",
    "cmap_day = mpl.cm.Spectral\n",
    "norm_day = mpl.colors.Normalize(vmin=-21, vmax=21)\n",
    "\n",
    "# Prior_Tx_binned, Prior_Tx_binned_colvec\n",
    "df_colvecs.loc[(df_colvecs[\"Prior_Tx\"] < 4) & (df_colvecs[\"Prior_Tx\"] >= 0), \"Prior_Tx_binned\"] = 1\n",
    "df_colvecs.loc[(df_colvecs[\"Prior_Tx\"] > 3) & (df_colvecs[\"Prior_Tx\"] < 7), \"Prior_Tx_binned\"] = 2\n",
    "df_colvecs.loc[(df_colvecs[\"Prior_Tx\"] > 6) & (df_colvecs[\"Prior_Tx\"] <= 10), \"Prior_Tx_binned\"] = 3\n",
    "df_colvecs[\"Prior_Tx_binned_colvec\"] = \"black\"\n",
    "df_colvecs.loc[df_colvecs.Prior_Tx_binned == 1,\"Prior_Tx_binned_colvec\"] = \"deepskyblue\"\n",
    "df_colvecs.loc[df_colvecs.Prior_Tx_binned == 2,\"Prior_Tx_binned_colvec\"] = \"gray\"\n",
    "df_colvecs.loc[df_colvecs.Prior_Tx_binned == 3,\"Prior_Tx_binned_colvec\"] = \"red\"\n",
    "colvec_label_Prior_Tx_binned   = ['0-3','4-6', '7-10', 'Not reported']\n",
    "colvec_color_Prior_Tx_binned  = [\"deepskyblue\", \"gray\",\"red\", \"black\"]\n",
    "colvec_marker_Prior_Tx_binned    = [\"o\",\"o\",\"o\", \"o\"]\n",
    "\n",
    "# CR+none-LR / nonCR+non-LR / CR+HR / nonCR+HR :\n",
    "df_colvecs.loc[(df_colvecs[\"CR180_binarized\"] == 1 ) & (df_colvecs[\"Basal_no_HighRisk_sample_binarized\"] == 1), \"CR180_noHRAbx\"] = \"noHR + CR\"\n",
    "df_colvecs.loc[(df_colvecs[\"CR180_binarized\"] == 0 ) & (df_colvecs[\"Basal_no_HighRisk_sample_binarized\"] == 1), \"CR180_noHRAbx\"] = \"noHR + noCR\"\n",
    "df_colvecs.loc[(df_colvecs[\"CR180_binarized\"] == 1 ) & (df_colvecs[\"Basal_no_HighRisk_sample_binarized\"] == 0), \"CR180_noHRAbx\"] = \"HR + CR\"\n",
    "df_colvecs.loc[(df_colvecs[\"CR180_binarized\"] == 0 ) & (df_colvecs[\"Basal_no_HighRisk_sample_binarized\"] == 0), \"CR180_noHRAbx\"] = \"HR + noCR\"\n",
    "\n",
    "# Subsettting df_colvecs:\n",
    "df_colvecs_all = df_colvecs.copy()\n",
    "\n",
    "df_CART_table_stool_basal_all_mean.index = df_CART_table_stool_basal_all_mean.stool_sample_id\n",
    "df_CART_table_stool_basal_NoHighRisk_mean.index = df_CART_table_stool_basal_NoHighRisk_mean.stool_sample_id\n",
    "\n",
    "df_colvecs_basal_all_mean = df_colvecs_all.loc[df_CART_table_stool_basal_all_mean.index,:]\n",
    "df_colvecs_basal_NoHighRisk_mean = df_colvecs.loc[ df_CART_table_stool_basal_NoHighRisk_mean.index,:]\n",
    "\n",
    "df_colvecs_basal_all_mean.index = df_CART_table_stool_basal_all_mean.subjectID\n",
    "df_colvecs_basal_NoHighRisk_mean.index = df_CART_table_stool_basal_NoHighRisk_mean.subjectID\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Beta-diversity PCoA analyses\n",
    "* Figure 2E (left and right) \n",
    "* Extended Data Figures 5E (left and right), 5F, 5G and 5H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MarkerHandler(HandlerBase):\n",
    "    def create_artists(self, legend, tup,xdescent, ydescent,\n",
    "                        width, height, fontsize,trans):\n",
    "        return [plt.Line2D([width/2], [height/2.],ls=\"\",\n",
    "                       marker=tup[1],color=tup[0], transform=trans)]\n",
    "    \n",
    "\n",
    "def plot_dr(df_X_raw, abundance_threshold, colvec, list_color = None ,list_mak = None, list_lab = None, title_name = None, cmap = None, norm = None, cbar = False):\n",
    "    \n",
    "    df_X = abundance_filtration(df_X_raw, abundance_threshold)\n",
    "    curr_figsize = [8,8]\n",
    "    \n",
    "    df_X_dist = distance.squareform(distance.pdist(df_X, metric=\"braycurtis\"))\n",
    "    df_X_pcoa = pcoa(df_X_dist)\n",
    "    pcoa_explained_var = np.ndarray.round(np.array(df_X_pcoa.proportion_explained[[\"PC1\", \"PC2\"]] * 100) ,1)\n",
    "\n",
    "    plt.figure(figsize = curr_figsize)\n",
    "    \n",
    "    pcoa_1 = np.array(df_X_pcoa.samples['PC1'])\n",
    "    pcoa_2 = np.array(df_X_pcoa.samples['PC2'])\n",
    "    \n",
    "    pcoa_1 = pcoa_1[~pd.isna( colvec)]\n",
    "    pcoa_2 = pcoa_2[~pd.isna( colvec)]\n",
    "    colvec = colvec[~pd.isna( colvec)]\n",
    "    \n",
    "    plt.scatter(pcoa_1, pcoa_2, c = colvec, cmap = cmap, norm = norm)\n",
    "    plt.title(\"PCoA_\"+ title_name,fontsize=16)\n",
    "    plt.xlabel('PCoA axis 1 ['+ np.array2string(pcoa_explained_var[0])+'%]',fontsize = 16)\n",
    "    plt.ylabel('PCoA axis 2 ['+ np.array2string(pcoa_explained_var[1])+'%]',fontsize = 16)\n",
    "    \n",
    "    if cbar == False:\n",
    "        plt.legend(list(zip(list_color,list_mak)), list_lab, handler_map={tuple:MarkerHandler()}) \n",
    "    \n",
    "    if cbar == True:   \n",
    "        plt.colorbar(mpl.cm.ScalarMappable(norm=norm, cmap=cmap))\n",
    "        \n",
    "    plt.savefig(path_output + title_name + '.pdf')\n",
    "    plt.close()\n",
    "    \n",
    "    return\n",
    "\n",
    "\n",
    "# Fig_2E (left and right)\n",
    "plot_dr(df_CART_table_taxa_all, taxa_abundance_threshold,  df_colvecs_all[\"AbxRisk_On_Sampling_colvec\"], colvec_color_AbxRisk_On_Sampling, colvec_marker_AbxRisk_On_Sampling, colvec_label_AbxRisk_On_Sampling, \"CART_Microbiome_Fig_2E_left_taxa\")\n",
    "plot_dr(df_CART_table_pathways_all, pathways_abundance_threshold,  df_colvecs_all[\"AbxRisk_On_Sampling_colvec\"], colvec_color_AbxRisk_On_Sampling, colvec_marker_AbxRisk_On_Sampling, colvec_label_AbxRisk_On_Sampling, \"CART_Microbiome_Fig_2E_right_pathwayds\")\n",
    "\n",
    "# Fig_S5E (left and right)\n",
    "plot_dr(df_CART_table_taxa_all, taxa_abundance_threshold,  df_colvecs_all[\"country_colvec\"], colvec_color_country, colvec_marker_country, colvec_label_country , \"CART_Microbiome_Fig_S5E_left_taxa\")\n",
    "plot_dr(df_CART_table_pathways_all, pathways_abundance_threshold,  df_colvecs_all[\"country_colvec\"], colvec_color_country, colvec_marker_country, colvec_label_country , \"CART_Microbiome_Fig_S5E_right_pathways\")\n",
    "\n",
    "# Fig_S5F\n",
    "plot_dr(df_CART_table_taxa_all, taxa_abundance_threshold,  df_colvecs_all[\"ECOG_01_colvec\"], colvec_color_ECOG_01, colvec_marker_ECOG_01, colvec_label_ECOG_01 , \"CART_Microbiome_Fig_S5F\")\n",
    "\n",
    "# Fig_S5G\n",
    "plot_dr(df_CART_table_taxa_all.loc[~np.isnan(df_colvecs_all[\"Prior_Tx_binned\"])], taxa_abundance_threshold,  df_colvecs_all.loc[~np.isnan(df_colvecs_all[\"Prior_Tx_binned\"]), \"Prior_Tx_binned_colvec\"], colvec_color_Prior_Tx_binned[0:3], colvec_marker_Prior_Tx_binned[0:3], colvec_label_Prior_Tx_binned[0:3] , \"CART_Microbiome_Fig_S5G\")\n",
    "\n",
    "# Fig_S5H\n",
    "plot_dr(df_CART_table_taxa_all, taxa_abundance_threshold,  df_colvecs_all[\"day_colvec\"], title_name= \"CART_Microbiome_Fig_S5H\", cmap = cmap_day, norm = norm_day, cbar=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential-abundance analyses\n",
    "* Figure 2F\n",
    "* Extended Figures 6A, 6B, 6C, 6D, 7A, 7B, 7C and 7D (volcano plots and corresponding log2 fold-change bar-plots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------------------------------------------------\n",
    "# Differential abundance analyses part 1: Defining the functions\n",
    "# ------------------------------------------------------------------------\n",
    "\n",
    "def differential_abundance_analysis(df_X_raw, abundance_threshold, df_y, split_label):\n",
    "  \n",
    "    df_X = abundance_filtration(df_X_raw, abundance_threshold)\n",
    "  \n",
    "    mannwhitneyu_results_df = pd.DataFrame(index = df_X.columns)\n",
    "    \n",
    "    df_X_no = df_X.loc[(df_y[split_label] == 0) | (df_y[split_label] == \"no\") ,:]\n",
    "    df_X_yes = df_X.loc[(df_y[split_label] == 1) | (df_y[split_label] == \"yes\"),:]\n",
    "    \n",
    "    for current_feature in df_X.columns:\n",
    "        current_feature_no = df_X_no[current_feature]\n",
    "        current_feature_yes = df_X_yes[current_feature]\n",
    "            \n",
    "        statistic, pvalue = mannwhitneyu(current_feature_no,current_feature_yes,alternative='two-sided')\n",
    "        \n",
    "        mannwhitneyu_results_df.loc[current_feature, \"log2_fold_change\"] = np.log2(current_feature_yes.mean() / current_feature_no.mean())\n",
    "        mannwhitneyu_results_df.loc[current_feature, \"Pval\"] = pvalue\n",
    "            \n",
    "    padj_binary, padj =  fdrcorrection(mannwhitneyu_results_df.Pval, alpha=0.2, method = 'i')\n",
    "    mannwhitneyu_results_df.loc[:, \"Padjusted\"] = padj\n",
    "    \n",
    "    return(mannwhitneyu_results_df)\n",
    "\n",
    "\n",
    "def plot_volcano(mannwhitneyu_res_df, out_data_path, title_name, p_thr,  label_yes = \"Yes\", label_no = \"No\"):\n",
    "    \n",
    "    fold_change_threshold = 2\n",
    "    \n",
    "    feature_names = list(mannwhitneyu_res_df.index)\n",
    "    log_fold_change = mannwhitneyu_res_df[\"log2_fold_change\"]\n",
    "    p_value = mannwhitneyu_res_df[\"Pval\"]\n",
    "    \n",
    "    hits_index = (p_value < p_thr) & (abs(log_fold_change) > np.log2(fold_change_threshold))\n",
    "    \n",
    "    # The volcano plot:\n",
    "    mpl.rcParams.update(mpl.rcParamsDefault)\n",
    "    plt.rcParams[\"font.size\"] = \"6\"\n",
    "    plt.figure(figsize=[10,10])\n",
    "    plt.scatter(log_fold_change, -np.log10(p_value),c='gray',s=8,label='Other features')\n",
    "    plt.scatter(log_fold_change[hits_index],-np.log10(p_value[hits_index]),c='red',s = 30, label='Hit features')\n",
    "    plt.xlabel('Log2 fold-change ('+ label_yes + \" / \" + label_no +\" )\",fontsize = 16)\n",
    "    plt.ylabel('-Log10 Pval',fontsize = 16)\n",
    "    plt.ylim(0,max(-np.log10(p_value) + 0.1))\n",
    "    plt.xlim(-max(abs(log_fold_change)) -0.2 ,max(abs(log_fold_change)) + 0.2)\n",
    "    plt.title(title_name,fontsize=16)\n",
    "    plt.plot([0,0], [0, max(-np.log10(p_value)) + 0.1], c = \"black\")\n",
    "    plt.plot([-np.log2(fold_change_threshold), -np.log2(fold_change_threshold)], [0, max(-np.log10(p_value)) + 0.1], '--',c = \"gray\")\n",
    "    plt.plot([np.log2(fold_change_threshold), np.log2(fold_change_threshold)], [0, max(-np.log10(p_value)) + 0.1], '--',c = \"gray\")\n",
    "    plt.plot([-max(abs(log_fold_change)) -0.2, max(abs(log_fold_change)) + 0.2], [-np.log10(p_thr), -np.log10(p_thr)], '--',c = \"gray\")\n",
    "    plt.legend(loc=(1.03,0.8),fontsize=14,fancybox=True)\n",
    "    plt.savefig(path_output + title_name + '_volcano.pdf',dpi=300, bbox_inches='tight')\n",
    "    plt.close()\n",
    " \n",
    "    # The associated bar-plot:\n",
    "    \n",
    "    mannwhitneyu_res_df_significant = mannwhitneyu_res_df.loc[(mannwhitneyu_res_df.Pval < p_thr) & \n",
    "                                                                                 (abs(mannwhitneyu_res_df.log2_fold_change) > np.log2(fold_change_threshold)),:]\n",
    "    if len(hits_index) > -1:\n",
    "\n",
    "        mannwhitneyu_res_df_significant_sorted = mannwhitneyu_res_df_significant.sort_values(by = \"log2_fold_change\")\n",
    "        \n",
    "        plt.figure(figsize=[10,20])\n",
    "        plt.barh(mannwhitneyu_res_df_significant_sorted.index, mannwhitneyu_res_df_significant_sorted.log2_fold_change)\n",
    "        \n",
    "        plt.xlabel('Log2 fold-change ('+ label_yes + \" / \" + label_no +\" )\",fontsize = 16)\n",
    "        plt.tick_params(axis='both', which='major', labelsize=10)\n",
    "        plt.savefig(path_output + title_name + '_bars.pdf',dpi=300, bbox_inches='tight')\n",
    "        plt.close()\n",
    "    \n",
    "    return\n",
    "\n",
    "\n",
    "def derive_all_differential_abundance_results(microbiome_df, microbiome_thr, col_vec_df,  subset_name, stratifeing_parameter, figfilename):\n",
    "    \n",
    "    df_res_dabund = differential_abundance_analysis(microbiome_df, microbiome_thr, col_vec_df, stratifeing_parameter)\n",
    "    combined_name = \"_\" + subset_name + \"_\" + stratifeing_parameter\n",
    "    plot_volcano(df_res_dabund, path_output, figfilename, p_thr = 0.05)\n",
    "\n",
    "    return(df_res_dabund)\n",
    "    \n",
    "\n",
    "# ------------------------------------------------------------------------\n",
    "# Differential abundance analyses part 2: Performing the analyses\n",
    "# ------------------------------------------------------------------------\n",
    "\n",
    "# Fig_2F\n",
    "df_res_dabund_basal_all_Basal_no_HighRisk_sample_taxa= \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_taxa_basal_all_mean,\\\n",
    "        taxa_abundance_threshold, df_colvecs_basal_all_mean, \\\n",
    "        \"Basal_all\", \"Basal_no_HighRisk_sample\", \"CART_Microbiome_Fig_2F\") \n",
    "\n",
    "\n",
    "# Fig_S6AB\n",
    "df_res_dabund_basal_all_Basal_no_HighRisk_sample_pathways = \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_pathways_basal_all_mean, \\\n",
    "        pathways_abundance_threshold, df_colvecs_basal_all_mean, \\\n",
    "        \"Basal_all\", \"Basal_no_HighRisk_sample\", \"CART_Microbiome_Fig_S6AB_pathways\") \n",
    "\n",
    "\n",
    "# Fig_S6CD (volcano) and S6D (bar plot) taxa\n",
    "df_res_dabund_basal_all_Country_taxa = \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_taxa_basal_all_mean,\\\n",
    "        taxa_abundance_threshold, df_colvecs_basal_all_mean, \\\n",
    "        \"Basal_all\", \"Country_yesno\", \"CART_Microbiome_Fig_S6CD_taxa\" ) \n",
    "\n",
    "\n",
    "# Fig_S6CD (volcano) and S6D (bar plot) pathways\n",
    "df_res_dabund_basal_all_Country_pathways = \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_pathways_basal_all_mean, \\\n",
    "        pathways_abundance_threshold, df_colvecs_basal_all_mean, \\\n",
    "        \"Basal_all\", \"Country_yesno\", \"CART_Microbiome_Fig_S6CD_pathways\" ) \n",
    "\n",
    "\n",
    "# Fig_S7A\n",
    "df_res_dabund_basal_all_Survival_taxa= \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_taxa_basal_all_mean, \\\n",
    "        taxa_abundance_threshold,  df_colvecs_basal_all_mean, \\\n",
    "        \"Basal_all\", \"Survival_d180\", \"CART_Microbiome_Fig_S7A\")\n",
    "\n",
    "\n",
    "# Fig_S7B\n",
    "df_res_dabund_basal_NoHighRisk_Survival_taxa = \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_taxa_basal_NoHighRisk_mean, \\\n",
    "        taxa_abundance_threshold, df_colvecs_basal_NoHighRisk_mean, \\\n",
    "        \"Basal_NoHighRisk\", \"Survival_d180\", \"CART_Microbiome_Fig_S7B\")\n",
    "\n",
    "\n",
    "# Fig_S7C\n",
    "df_res_dabund_basal_NoHighRisk_Survival_pathways = \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_pathways_basal_NoHighRisk_mean, \\\n",
    "        pathways_abundance_threshold, df_colvecs_basal_NoHighRisk_mean, \\\n",
    "        \"Basal_NoHighRisk\", \"Survival_d180\", \"CART_Microbiome_Fig_S7C\")\n",
    "\n",
    "\n",
    "# Fig_S7D\n",
    "df_res_dabund_basal_NoHighRisk_Progression_early_pathways = \\\n",
    "        derive_all_differential_abundance_results( \\\n",
    "        df_CART_table_pathways_basal_NoHighRisk_mean, \\\n",
    "        pathways_abundance_threshold, df_colvecs_basal_all_mean, \\\n",
    "        \"Basal_NoHighRisk\", \"Progression_early\", \"CART_Microbiome_Fig_S7D\") \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PERMANOVA analyses\n",
    "* Figures 2D, 2G, 2H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------------------------------------------------\n",
    "# PERMANOVA part 1: Performing the PERMANOVA analyses\n",
    "# ------------------------------------------------------------------------\n",
    "\n",
    "def analyze_permanova(df_X, grouping_vec, abundance_threshold):\n",
    "    np.random.seed(0)\n",
    "\n",
    "    df_X_filtered = abundance_filtration(df_X, abundance_threshold)\n",
    "    \n",
    "    df_X_filtered_noNaN = df_X_filtered.loc[pd.isna(grouping_vec) == False,:]\n",
    "    grouping_vec_noNaN = grouping_vec[pd.isna(grouping_vec) == False]\n",
    "    \n",
    "    grouping_vec_noNaN = grouping_vec_noNaN.astype(str)\n",
    "    \n",
    "    dm_X = DistanceMatrix(distance.squareform(distance.pdist(df_X_filtered_noNaN, metric='braycurtis'))) # PERMANOVA based on braycurtis (addded on 06 June 2022)\n",
    "\n",
    "    permanova_results = permanova(dm_X, grouping_vec_noNaN)\n",
    "\n",
    "    r2 = 1 - 1 / (1 + permanova_results[4] * permanova_results[3] / (permanova_results[2] - permanova_results[3] - 1))\n",
    "    permanova_results[\"R2\"] = r2\n",
    "    \n",
    "    return(permanova_results)\n",
    "\n",
    "\n",
    "permanova_cateogry_names = [\"Response (day 180)\", \\\n",
    "                            \"Survival (day 180)\", \\\n",
    "                            \"Early progression\", \\\n",
    "                            \"CRS > 1\", \\\n",
    "                            \"ICANS grades\",\\\n",
    "                            \"ECOG > 1\", \\\n",
    "                            \"no_HighRisk_sample\", \\\n",
    "                            \"Country\", \\\n",
    "                            \"Gender\", \\\n",
    "                            \"Age >= 65\", \\\n",
    "                            \"LDH depletion >= 280\", \\\n",
    "                            \"Bulky disease 10cm\", \\\n",
    "                            \"Prior therapies 0-3 / 4-6 / 7-10\",\n",
    "                            \"NHR_Response\",\n",
    "                            \"NHS_Survival\",\n",
    "                            \"NHR_Progression\",\n",
    "                            \"NHR_CRS\",\n",
    "                            \"NHR_ICANS\"]\n",
    "\n",
    "\n",
    "# PERMANOVA analyses based on all the basal samples:\n",
    "# --------------------------------------------------\n",
    "\n",
    "# Criterion: Response (day 180) (i.e. Complete remission, CR, at day 180)\n",
    "permanova_taxa_basal_CR180 = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.CR180_binarized, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_CR180 = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.CR180_binarized, pathways_abundance_threshold)\n",
    "permanova_genes_basal_CR180 = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.CR180_binarized, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Survival (day 180)\n",
    "permanova_taxa_basal_Survival = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Survival_d180, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_Survival = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Survival_d180, pathways_abundance_threshold)\n",
    "permanova_genes_basal_Survival = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Survival_d180, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Early progression\n",
    "permanova_taxa_basal_Progression_early = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Progression_early_binarized, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_Progression_early = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Progression_early_binarized, pathways_abundance_threshold)\n",
    "permanova_genes_basal_Progression_early = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Progression_early_binarized, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: CRS > 1\n",
    "permanova_taxa_basal_CRS_01 = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.CRS_01, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_CRS_01 = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.CRS_01, pathways_abundance_threshold)\n",
    "permanova_genes_basal_CRS_01 = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.CRS_01, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: ICANS grades\n",
    "permanova_taxa_basal_ICANS_grade = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.ICANS_grade, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_ICANS_grade = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.ICANS_grade, pathways_abundance_threshold)\n",
    "permanova_genes_basal_ICANS_grade = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.ICANS_grade, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: ECOG > 1\n",
    "permanova_taxa_basal_ECOG_01 = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.ECOG_01, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_ECOG_01 = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.ECOG_01, pathways_abundance_threshold)\n",
    "permanova_genes_basal_ECOG_01 = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.ECOG_01, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: no_HighRisk_sample\n",
    "permanova_taxa_basal_no_HighRisk_sample = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Basal_no_HighRisk_sample, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_no_HighRisk_sample = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Basal_no_HighRisk_sample, pathways_abundance_threshold)\n",
    "permanova_genes_basal_no_HighRisk_sample = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Basal_no_HighRisk_sample, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Country\n",
    "permanova_taxa_basal_Country = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Country, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_Country = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Country, pathways_abundance_threshold)\n",
    "permanova_genes_basal_Country = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Country, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Gender\n",
    "permanova_taxa_basal_Gender = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Gender, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_Gender = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Gender, pathways_abundance_threshold)\n",
    "permanova_genes_basal_Gender = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Gender, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Age >= 65\n",
    "permanova_taxa_basal_Age = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Age_binarized, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_Age = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Age_binarized, pathways_abundance_threshold)\n",
    "permanova_genes_basal_Age = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Age_binarized, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: LDH depletion >= 280\n",
    "permanova_taxa_basal_LDH_Depletion = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.LDH_Depletion_binarized, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_LDH_Depletion = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.LDH_Depletion_binarized, pathways_abundance_threshold)\n",
    "permanova_genes_basal_LDH_Depletion = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.LDH_Depletion_binarized, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: BulkyDisease_10cm\n",
    "permanova_taxa_basal_BulkyDisease = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.BulkyDisease_10cm, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_BulkyDisease = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.BulkyDisease_10cm, pathways_abundance_threshold)\n",
    "permanova_genes_basal_BulkyDisease = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.BulkyDisease_10cm, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Prior therapies 0-3 / 4-6 / 7-10\n",
    "permanova_taxa_basal_Prior_Tx_binned = analyze_permanova(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean.Prior_Tx_binned, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_Prior_Tx_binned = analyze_permanova(df_CART_table_pathways_basal_all_mean, df_colvecs_basal_all_mean.Prior_Tx_binned, pathways_abundance_threshold)\n",
    "permanova_genes_basal_Prior_Tx_binned = analyze_permanova(df_CART_table_genes_basal_all_mean, df_colvecs_basal_all_mean.Prior_Tx_binned, genes_abundance_threshold)\n",
    "\n",
    "\n",
    "# PERMANOVA analyses based on the no_HighRisk basal samples:\n",
    "# ----------------------------------------------------------\n",
    "\n",
    "# Criterion: Response (day 180) (i.e. Complete remission, CR, at day 180)\n",
    "permanova_taxa_basal_NoHighRisk_CR180 = analyze_permanova(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.CR180_binarized, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_NoHighRisk_CR180 = analyze_permanova(df_CART_table_pathways_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.CR180_binarized, pathways_abundance_threshold)\n",
    "permanova_genes_basal_NoHighRisk_CR180 = analyze_permanova(df_CART_table_genes_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.CR180_binarized, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Survival (day 180)\n",
    "permanova_taxa_basal_NoHighRisk_Survival = analyze_permanova(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.Survival_d180, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_NoHighRisk_Survival = analyze_permanova(df_CART_table_pathways_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.Survival_d180, pathways_abundance_threshold)\n",
    "permanova_genes_basal_NoHighRisk_Survival = analyze_permanova(df_CART_table_genes_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.Survival_d180, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: Early progression\n",
    "permanova_taxa_basal_NoHighRisk_Progression_early = analyze_permanova(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.Progression_early_binarized, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_NoHighRisk_Progression_early = analyze_permanova(df_CART_table_pathways_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.Progression_early_binarized, pathways_abundance_threshold)\n",
    "permanova_genes_basal_NoHighRisk_Progression_early = analyze_permanova(df_CART_table_genes_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.Progression_early_binarized, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: CRS > 1\n",
    "permanova_taxa_basal_NoHighRisk_CRS_01 = analyze_permanova(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.CRS_01, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_NoHighRisk_CRS_01 = analyze_permanova(df_CART_table_pathways_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.CRS_01, pathways_abundance_threshold)\n",
    "permanova_genes_basal_NoHighRisk_CRS_01 = analyze_permanova(df_CART_table_genes_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.CRS_01, genes_abundance_threshold)\n",
    "\n",
    "# Criterion: ICANS grades\n",
    "permanova_taxa_basal_NoHighRisk_ICANS_grade = analyze_permanova(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.ICANS_grade, taxa_abundance_threshold)\n",
    "permanova_pathways_basal_NoHighRisk_ICANS_grade = analyze_permanova(df_CART_table_pathways_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.ICANS_grade, pathways_abundance_threshold)\n",
    "permanova_genes_basal_NoHighRisk_ICANS_grade = analyze_permanova(df_CART_table_genes_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean.ICANS_grade, genes_abundance_threshold)\n",
    "\n",
    "# ------------------------------------------------------------------------\n",
    "# PERMANOVA part 2: Organizing the PERMANOVA results as heatmaps\n",
    "# ------------------------------------------------------------------------\n",
    "\n",
    "# Taxa permanova results:\n",
    "\n",
    "permanova_taxa_list =[permanova_taxa_basal_CR180,\n",
    "                      permanova_taxa_basal_Survival,\n",
    "                      permanova_taxa_basal_Progression_early,\n",
    "                      permanova_taxa_basal_CRS_01,\n",
    "                      permanova_taxa_basal_ICANS_grade,\n",
    "                      permanova_taxa_basal_ECOG_01,\n",
    "                      permanova_taxa_basal_no_HighRisk_sample,\n",
    "                      permanova_taxa_basal_Country,\n",
    "                      permanova_taxa_basal_Gender,\n",
    "                      permanova_taxa_basal_Age,\n",
    "                      permanova_taxa_basal_LDH_Depletion,\n",
    "                      permanova_taxa_basal_BulkyDisease,\n",
    "                      permanova_taxa_basal_Prior_Tx_binned,\n",
    "                      permanova_taxa_basal_NoHighRisk_CR180,\n",
    "                      permanova_taxa_basal_NoHighRisk_Survival,\n",
    "                      permanova_taxa_basal_NoHighRisk_Progression_early,\n",
    "                      permanova_taxa_basal_NoHighRisk_CRS_01,\n",
    "                      permanova_taxa_basal_NoHighRisk_ICANS_grade]\n",
    "\n",
    "permanova_taxa_basal_all =  pd.concat(permanova_taxa_list, axis=1)\n",
    "permanova_taxa_basal_all.columns = permanova_cateogry_names\n",
    "\n",
    "\n",
    "# Pathways permanova results:\n",
    "\n",
    "permanova_pathways_list =[permanova_pathways_basal_CR180,\n",
    "                      permanova_pathways_basal_Survival,\n",
    "                      permanova_pathways_basal_Progression_early,\n",
    "                      permanova_pathways_basal_CRS_01,\n",
    "                      permanova_pathways_basal_ICANS_grade,\n",
    "                      permanova_pathways_basal_ECOG_01,\n",
    "                      permanova_pathways_basal_no_HighRisk_sample,\n",
    "                      permanova_pathways_basal_Country,\n",
    "                      permanova_pathways_basal_Gender,\n",
    "                      permanova_pathways_basal_Age,\n",
    "                      permanova_pathways_basal_LDH_Depletion,\n",
    "                      permanova_pathways_basal_BulkyDisease,\n",
    "                      permanova_pathways_basal_Prior_Tx_binned,\n",
    "                      permanova_pathways_basal_NoHighRisk_CR180,\n",
    "                      permanova_pathways_basal_NoHighRisk_Survival,\n",
    "                      permanova_pathways_basal_NoHighRisk_Progression_early,\n",
    "                      permanova_pathways_basal_NoHighRisk_CRS_01,\n",
    "                      permanova_pathways_basal_NoHighRisk_ICANS_grade]\n",
    "\n",
    "permanova_pathways_basal_all =  pd.concat(permanova_pathways_list, axis=1)\n",
    "permanova_pathways_basal_all.columns = permanova_cateogry_names\n",
    "\n",
    "\n",
    "# Genes permanova results:\n",
    "\n",
    "permanova_genes_list =[permanova_genes_basal_CR180,\n",
    "                      permanova_genes_basal_Survival,\n",
    "                      permanova_genes_basal_Progression_early,\n",
    "                      permanova_genes_basal_CRS_01,\n",
    "                      permanova_genes_basal_ICANS_grade,\n",
    "                      permanova_genes_basal_ECOG_01,\n",
    "                      permanova_genes_basal_no_HighRisk_sample,\n",
    "                      permanova_genes_basal_Country,\n",
    "                      permanova_genes_basal_Gender,\n",
    "                      permanova_genes_basal_Age,\n",
    "                      permanova_genes_basal_LDH_Depletion,\n",
    "                      permanova_genes_basal_BulkyDisease,\n",
    "                      permanova_genes_basal_Prior_Tx_binned,\n",
    "                      permanova_genes_basal_NoHighRisk_CR180,\n",
    "                      permanova_genes_basal_NoHighRisk_Survival,\n",
    "                      permanova_genes_basal_NoHighRisk_Progression_early,\n",
    "                      permanova_genes_basal_NoHighRisk_CRS_01,\n",
    "                      permanova_genes_basal_NoHighRisk_ICANS_grade]\n",
    "\n",
    "\n",
    "permanova_genes_basal_all =  pd.concat(permanova_genes_list, axis=1)\n",
    "permanova_genes_basal_all.columns = permanova_cateogry_names\n",
    "\n",
    "permanova_basal_summary_pval = pd.DataFrame(index = permanova_taxa_basal_all.columns, columns=[\"Species\", \"Pathways\", \"Genes\"] )\n",
    "permanova_basal_summary_pval.loc[:,\"Species\"] = permanova_taxa_basal_all.loc[\"p-value\",:]\n",
    "permanova_basal_summary_pval.loc[:,\"Pathways\"] = permanova_pathways_basal_all.loc[\"p-value\",:]\n",
    "permanova_basal_summary_pval.loc[:,\"Genes\"] = permanova_genes_basal_all.loc[\"p-value\",:]\n",
    "\n",
    "permanova_basal_summary_pval_symbol = permanova_basal_summary_pval.copy()\n",
    "\n",
    "permanova_basal_summary_pval_symbol[permanova_basal_summary_pval >= 0.05] = \"\"\n",
    "permanova_basal_summary_pval_symbol[permanova_basal_summary_pval < 0.05] = \"\\n*\"\n",
    "permanova_basal_summary_pval_symbol[permanova_basal_summary_pval < 0.01] = \"\\n**\"\n",
    "permanova_basal_summary_pval_symbol[permanova_basal_summary_pval < 0.001] = \"\\n***\"\n",
    "\n",
    "permanova_basal_summary_R2 = pd.DataFrame(index = permanova_taxa_basal_all.columns, columns=[\"Species\", \"Pathways\", \"Genes\"] )\n",
    "permanova_basal_summary_R2.loc[:,\"Species\"] = permanova_taxa_basal_all.loc[\"R2\",:]\n",
    "permanova_basal_summary_R2.loc[:,\"Pathways\"] = permanova_pathways_basal_all.loc[\"R2\",:]\n",
    "permanova_basal_summary_R2.loc[:,\"Genes\"] = permanova_genes_basal_all.loc[\"R2\",:]\n",
    "permanova_basal_summary_R2_prercentage = round((permanova_basal_summary_R2.astype(\"float\") * 100),1)\n",
    "\n",
    "\n",
    "# Fig_2D:\n",
    "\n",
    "permanova_table_1_categogeries = [\"Age >= 65\", \"Gender\",  \"Country\", \"ECOG > 1\", \"LDH depletion >= 280\", \"Bulky disease 10cm\", \"Prior therapies 0-3 / 4-6 / 7-10\", \"no_HighRisk_sample\"]\n",
    "permanova_table_1_category_lables = [\"Age\", \"Gender\", \"Country\", \"ECOG\", \"LDH\", \"Bulky disease\", \"Prior therapies\", \"Antibiotics\"]\n",
    "\n",
    "permanova_table_1_R2 = permanova_basal_summary_R2_prercentage.loc[permanova_table_1_categogeries,:]\n",
    "permanova_table_1_R2.index = permanova_table_1_category_lables\n",
    "permanova_table_1_pval_symbols = permanova_basal_summary_pval_symbol.loc[permanova_table_1_categogeries,:]\n",
    "permanova_table_1_labels =  permanova_table_1_R2.values.astype(str) + np.asarray(permanova_table_1_pval_symbols) \n",
    "\n",
    "plt.figure(figsize = (8,8))\n",
    "sns.set(font_scale=1.5)\n",
    "annot_kws = {\"verticalalignment\": 'center_baseline'}\n",
    "permanova_basa_plot_1 = sns.heatmap(permanova_table_1_R2, fmt=\"\", cmap='YlGnBu', annot = permanova_table_1_labels, linewidths=2, cbar_kws={'label': 'Variance explained (%)'}, annot_kws=annot_kws)\n",
    "permanova_basa_plot_1.set_xticklabels(permanova_basa_plot_1.get_xticklabels(), rotation=45, rotation_mode='anchor', ha='right')\n",
    "permanova_basa_plot_1 = permanova_basa_plot_1.get_figure()\n",
    "plt.tight_layout()\n",
    "permanova_basa_plot_1.savefig(path_output + 'CART_Microbiome_Fig_2D.pdf')\n",
    "plt.close()\n",
    "\n",
    "\n",
    "# Fig_2G:\n",
    "\n",
    "permanova_table_2_categogeries = [\"Response (day 180)\", \"Survival (day 180)\", \"Early progression\", \"CRS > 1\", \"ICANS grades\"]\n",
    "permanova_table_2_3_category_lables = [\"Response\", \"Survival\", \"Early progres.\", \"CRS\", \"ICANS\"]\n",
    "\n",
    "permanova_table_2_R2 = permanova_basal_summary_R2_prercentage.loc[permanova_table_2_categogeries,:]\n",
    "permanova_table_2_R2.index = permanova_table_2_3_category_lables\n",
    "permanova_table_2_pval_symbols = permanova_basal_summary_pval_symbol.loc[permanova_table_2_categogeries,:]\n",
    "permanova_table_2_labels =  permanova_table_2_R2.values.astype(str) + np.asarray(permanova_table_2_pval_symbols) \n",
    "plt.close()\n",
    "\n",
    "plt.figure(figsize = (8,6))\n",
    "sns.set(font_scale=1.5)\n",
    "permanova_basa_plot_2 = sns.heatmap(permanova_table_2_R2, fmt=\"\", cmap='YlGnBu', annot = permanova_table_2_labels, linewidths=2, cbar_kws={'label': 'Variance explained (%)'})\n",
    "permanova_basa_plot_2.set_xticklabels(permanova_basa_plot_2.get_xticklabels(), rotation=45, rotation_mode='anchor', ha='right')\n",
    "plt.title(\"All specimens\")\n",
    "permanova_basa_plot_2 = permanova_basa_plot_2.get_figure()\n",
    "plt.tight_layout()\n",
    "permanova_basa_plot_2.savefig(path_output + 'CART_Microbiome_Fig_2G.pdf')\n",
    "plt.close()\n",
    "\n",
    "\n",
    "# Fig_2H:\n",
    "\n",
    "permanova_table_3_categogeries = [\"NHR_Response\", \"NHS_Survival\", \"NHR_Progression\", \"NHR_CRS\", \"NHR_ICANS\"]\n",
    "permanova_table_3_R2 = permanova_basal_summary_R2_prercentage.loc[permanova_table_3_categogeries,:]\n",
    "permanova_table_3_R2.index = permanova_table_2_3_category_lables\n",
    "\n",
    "permanova_table_3_pval_symbols = permanova_basal_summary_pval_symbol.loc[permanova_table_3_categogeries,:]\n",
    "permanova_table_3_labels =  permanova_table_3_R2.values.astype(str) + np.asarray(permanova_table_3_pval_symbols) \n",
    "plt.close()\n",
    "\n",
    "plt.figure(figsize = (8,6))\n",
    "sns.set(font_scale=1.5)\n",
    "permanova_basa_plot_3 = sns.heatmap(permanova_table_3_R2, fmt=\"\", cmap='YlGnBu', annot = permanova_table_3_labels, linewidths=1.5, cbar_kws={'label': 'Variance explained (%)'})\n",
    "permanova_basa_plot_3.set_xticklabels(permanova_basa_plot_3.get_xticklabels(), rotation=45, rotation_mode='anchor', ha='right')\n",
    "plt.title(\"w/o HR abx samples\")\n",
    "permanova_basa_plot_3 = permanova_basa_plot_3.get_figure()\n",
    "plt.tight_layout()\n",
    "permanova_basa_plot_3.savefig(path_output + 'CART_Microbiome_Fig_2H.pdf')\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Heatmaps of associations between microbiome and clinical features\n",
    "* Extended Figures 8A, 8B, 9A and 9B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scan_associations(df_X_raw, network_associations_targets_df, abundance_threshold, pval_threshold, foldchange_threshold):\n",
    "    \n",
    "    df_X = abundance_filtration(df_X_raw, abundance_threshold)\n",
    "    \n",
    "    df_associations_network_pval = pd.DataFrame(index = df_X.columns)\n",
    "    \n",
    "    df_associations_network_change = df_associations_network_pval.copy()\n",
    "    df_associations_network_masked = df_associations_network_pval.copy()\n",
    "\n",
    "    for current_target in network_associations_targets_df.columns:\n",
    "        \n",
    "        current_microbiome_feature_0 = df_X.loc[(network_associations_targets_df[current_target] == 0),:]\n",
    "        current_microbiome_feature_1 = df_X.loc[(network_associations_targets_df[current_target] == 1),:]\n",
    "    \n",
    "        for current_microbiome_feature in df_X.columns:\n",
    "            \n",
    "            current_feature_no = current_microbiome_feature_0[current_microbiome_feature]\n",
    "            current_feature_yes = current_microbiome_feature_1[current_microbiome_feature]   \n",
    "            \n",
    "            statistic, pvalue = mannwhitneyu(current_feature_no,current_feature_yes,alternative='two-sided')\n",
    "            log10_fold_change = np.log10(current_feature_yes.mean() / current_feature_no.mean())\n",
    "                \n",
    "            df_associations_network_pval.loc[current_microbiome_feature, current_target] = pvalue\n",
    "            df_associations_network_change.loc[current_microbiome_feature, current_target] = log10_fold_change\n",
    "                \n",
    "            if ( (pvalue > pval_threshold) or (abs(log10_fold_change)) < np.log10(foldchange_threshold)):\n",
    "                df_associations_network_masked.loc[current_microbiome_feature, current_target] = 0\n",
    "            else:\n",
    "                df_associations_network_masked.loc[current_microbiome_feature, current_target] = log10_fold_change\n",
    "                               \n",
    "                \n",
    "    df_associations_network_masked = df_associations_network_masked.fillna(0)\n",
    "    df_associations_network_masked_abs = abs(df_associations_network_masked)\n",
    "    df_associations_network_masked_filtered = df_associations_network_masked.loc[df_associations_network_masked_abs.sum(axis = 1) > 0,:]\n",
    "               \n",
    "    return(df_associations_network_masked_filtered)\n",
    "\n",
    "\n",
    "network_associations_targets_df = pd.DataFrame(index = df_CART_table_stool_all.index)\n",
    "\n",
    "network_associations_targets_df[\"Age\"] = df_colvecs[\"Age_binarized\"]\n",
    "network_associations_targets_df[\"Country\"] = df_colvecs[\"Country_yesno\"]\n",
    "network_associations_targets_df[\"ECOG\"] = df_colvecs[\"ECOG_01\"]\n",
    "network_associations_targets_df[\"LDH\"] = df_colvecs[\"LDH_Depletion_binarized\"]\n",
    "network_associations_targets_df[\"Abx.\"] = df_colvecs[\"Basal_no_HighRisk_sample\"]\n",
    "network_associations_targets_df[\"CR\"] = df_colvecs[\"CR180_binarized\"]\n",
    "network_associations_targets_df[\"Survival\"] = df_colvecs[\"Survival_d180\"]\n",
    "network_associations_targets_df[\"Progression\"] = df_colvecs[\"Progression_early_binarized\"]\n",
    "network_associations_targets_df[\"CRS\"] = df_colvecs[\"CRS_01\"]\n",
    "\n",
    "network_associations_targets_df = network_associations_targets_df.replace(\"no\", 0)\n",
    "network_associations_targets_df = network_associations_targets_df.replace(\"yes\", 1)\n",
    "\n",
    "network_associations_targets_df = network_associations_targets_df.replace(False, 0)\n",
    "network_associations_targets_df = network_associations_targets_df.replace(True, 1)\n",
    "\n",
    "network_associations_targets_df_basal_all = network_associations_targets_df.loc[df_CART_table_stool_basal_all_mean.index,:]\n",
    "network_associations_targets_df_basal_NoHighRisk = network_associations_targets_df.loc[df_CART_table_stool_basal_NoHighRisk_mean.index,:]\n",
    "\n",
    "network_associations_targets_df_basal_NoHighRisk = network_associations_targets_df_basal_NoHighRisk.drop(columns='Abx.')\n",
    "\n",
    "network_associations_targets_df_basal_all.index = df_CART_table_stool_basal_all_mean.subjectID\n",
    "network_associations_targets_df_basal_NoHighRisk.index = df_CART_table_stool_basal_NoHighRisk_mean.subjectID\n",
    "\n",
    "\n",
    "# Fig_S8A:\n",
    "scan_associations_results_taxa_basal_005 = scan_associations(df_CART_table_taxa_basal_all_mean, network_associations_targets_df_basal_all, abundance_threshold = taxa_abundance_threshold, pval_threshold = 0.05, foldchange_threshold = 2)\n",
    "associations_heatmap_taxa_basal_005 = sns.clustermap(scan_associations_results_taxa_basal_005, cmap = \"vlag\", vmin = -1, vmax = 1, figsize = (20,80),  col_cluster = False)\n",
    "associations_heatmap_taxa_basal_005.savefig(path_output + 'CART_Microbiome_Fig_S8A.pdf')\n",
    "plt.close()\n",
    "\n",
    "# Fig_S8B:\n",
    "scan_associations_results_pathways_basal_005 = scan_associations(df_CART_table_pathways_basal_all_mean, network_associations_targets_df_basal_all, abundance_threshold = pathways_abundance_threshold, pval_threshold = 0.05, foldchange_threshold = 2)\n",
    "associations_heatmap_pathways_basal_005 = sns.clustermap(scan_associations_results_pathways_basal_005, cmap = \"vlag\", vmin = -1, vmax = 1, figsize = (20,80),  col_cluster = False)\n",
    "associations_heatmap_pathways_basal_005.savefig(path_output + 'CART_Microbiome_Fig_S8B.pdf')\n",
    "plt.close()\n",
    "\n",
    "# Fig_S9A:\n",
    "scan_associations_results_taxa_basal_005 = scan_associations(df_CART_table_taxa_basal_NoHighRisk_mean, network_associations_targets_df_basal_NoHighRisk, abundance_threshold = taxa_abundance_threshold, pval_threshold = 0.05, foldchange_threshold = 2)\n",
    "associations_heatmap_taxa_basal_005 = sns.clustermap(scan_associations_results_taxa_basal_005, cmap = \"vlag\", vmin = -1, vmax = 1, figsize = (20,80),  col_cluster = False)\n",
    "associations_heatmap_taxa_basal_005.savefig(path_output + 'CART_Microbiome_Fig_S9A.pdf')\n",
    "plt.close()\n",
    "\n",
    "# Fig_S9B:\n",
    "scan_associations_results_pathways_basal_005 = scan_associations(df_CART_table_pathways_basal_NoHighRisk_mean, network_associations_targets_df_basal_NoHighRisk, abundance_threshold = pathways_abundance_threshold, pval_threshold = 0.05, foldchange_threshold = 2)\n",
    "associations_heatmap_pathways_basal_005 = sns.clustermap(scan_associations_results_pathways_basal_005, cmap = \"vlag\", vmin = -1, vmax = 1, figsize = (20,80),  col_cluster = False)\n",
    "associations_heatmap_pathways_basal_005.savefig(path_output + 'CART_Microbiome_Fig_S9B.pdf')\n",
    "plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processing the ML-derived feature coefficients table as derived for CR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CR_coefficients_df_filtered = CR_coefficients_df.loc[abs(CR_coefficients_df.Coefficient) > 0,:].copy()\n",
    "CR_coefficients_df_filtered[\"Coefficient_absolute\"] = abs(CR_coefficients_df_filtered.Coefficient)\n",
    "\n",
    "features_perc_thr_CR = 25\n",
    "\n",
    "CR_total_above_perc_thr = round(CR_coefficients_df_filtered.shape[0] * (features_perc_thr_CR  / 100))\n",
    "CR_coefficients_df_top_perc = CR_coefficients_df_filtered.nlargest(CR_total_above_perc_thr,\"Coefficient_absolute\")\n",
    "CR_coefficients_df_top_perc = CR_coefficients_df_top_perc.sort_values(by = \"Coefficient\")\n",
    "\n",
    "CR_coefficients_df_top_perc.to_excel(path_output+\"CR_coefficients.xlsx\", index = False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identifying taxa associations with CD3, CD4, and CD8 T cells blood-levels at apharesis\n",
    "* Figures 4A and 4B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "apharesis_parameters = ['CD3_mul_Aph', 'CD4_mul_Aph', 'CD8_mul_Aph']\n",
    "\n",
    "def scan_correlation_associations(df_microbiome, df_Aph, microbiome_features, Aph_parameters, title_name):\n",
    "    \n",
    "    if microbiome_features == -1:\n",
    "        microbiome_features = df_microbiome.columns\n",
    "    \n",
    "    df_all_associations = pd.DataFrame(columns=[\"microbiome_feature\", \"Aph_parameter\", \"tau\", \"pval\"])\n",
    "\n",
    "    for current_Aph_parameter in Aph_parameters: # x\n",
    "        for current_microbiome_feature in microbiome_features: # y\n",
    "        \n",
    "            curr_x_all = df_Aph[current_Aph_parameter]\n",
    "            curr_y_all = df_microbiome[current_microbiome_feature]\n",
    "            curr_x = curr_x_all[(np.isnan(curr_x_all.values) == False) & (np.isnan(curr_y_all.values)== False)]\n",
    "            curr_y = curr_y_all[(np.isnan(curr_x_all.values) == False) & (np.isnan(curr_y_all.values)== False)]\n",
    "                        \n",
    "            rho, pval  = stats.kendalltau(curr_x.values , curr_y.values )\n",
    "       \n",
    "            current_index = current_Aph_parameter + \"---\"+current_microbiome_feature\n",
    "            df_all_associations.loc[current_index, \"Aph_parameter\"] = current_Aph_parameter\n",
    "            df_all_associations.loc[current_index, \"microbiome_feature\"] = current_microbiome_feature\n",
    "            df_all_associations.loc[current_index, \"tau\"] = rho\n",
    "            df_all_associations.loc[current_index, \"pval\"] = pval\n",
    "    \n",
    "    padj_binary, padj =  fdrcorrection(df_all_associations.pval, alpha=0.2, method = 'i')\n",
    "    \n",
    "    df_all_associations[\"Padj\"] = padj\n",
    "    \n",
    "    df_all_associations_filtered = df_all_associations.loc[df_all_associations.pval < 0.01]\n",
    "    df_all_associations_sorted = df_all_associations_filtered.sort_values(by = \"tau\")\n",
    "    \n",
    "    df_all_associations_sorted[\"pvalues_color\"] = df_all_associations_sorted.pval\n",
    "    df_all_associations_sorted.loc[df_all_associations_sorted.pval < 0.01, \"pvalues_color\"] = \"green\"\n",
    "    df_all_associations_sorted.loc[df_all_associations_sorted.Padj < 0.05, \"pvalues_color\"] = \"blue\"\n",
    "    df_all_associations_sorted.loc[df_all_associations_sorted.Padj < 0.01,  \"pvalues_color\"] = \"red\"\n",
    "    \n",
    "    plt.close()\n",
    "    plt.barh(y = df_all_associations_sorted.index , width = df_all_associations_sorted.tau, color = df_all_associations_sorted[\"pvalues_color\"])\n",
    "    plt.xlabel(\"tau\")\n",
    "    plt.title(title_name)\n",
    "    plt.savefig(path_output + title_name+\".pdf\",dpi=300, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    \n",
    "    return(df_all_associations_sorted)\n",
    "\n",
    "mpl.rcParams.update(mpl.rcParamsDefault)\n",
    "plt.close()\n",
    "\n",
    "\n",
    "df_mul_Aph = pd.DataFrame(index=df_CART_table_stool_all.index)\n",
    "df_mul_Aph[\"subjectID\"] = df_CART_table_stool_all[\"subjectID\"] \n",
    "\n",
    "for current_patient in df_mul_Aph.subjectID:\n",
    "    df_mul_Aph.loc[df_mul_Aph[\"subjectID\"] == current_patient, \"CD3_mul_Aph\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CD3_mul_Aph\"].values\n",
    "    df_mul_Aph.loc[df_mul_Aph[\"subjectID\"] == current_patient, \"CD4_mul_Aph\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CD4_mul_Aph\"].values \n",
    "    df_mul_Aph.loc[df_mul_Aph[\"subjectID\"] == current_patient, \"CD8_mul_Aph\"] = df_CART_table_patients_filtered.loc[(df_CART_table_patients_filtered[\"subjectID\"] == current_patient) ,\"CD8_mul_Aph\"].values\n",
    "\n",
    "df_mul_Aph_basal_all_mean = df_mul_Aph.loc[df_CART_table_stool_basal_all_mean.index,:]\n",
    "df_mul_Aph_basal_NoHighRisk_mean = df_mul_Aph.loc[ df_CART_table_stool_basal_NoHighRisk_mean.index,:]\n",
    "\n",
    "df_mul_Aph_basal_all_mean.index = df_CART_table_stool_basal_all_mean.subjectID\n",
    "df_mul_Aph_basal_NoHighRisk_mean.index = df_CART_table_stool_basal_NoHighRisk_mean.subjectID\n",
    "\n",
    "\n",
    "# Fig_4A:\n",
    "df_apharesis_basal_all_taxa_ML_allTaxa = scan_correlation_associations(df_CART_table_taxa_basal_all_mean, df_mul_Aph_basal_all_mean, -1, apharesis_parameters, \"CART_Microbiome_Fig_4A\" )\n",
    "\n",
    "\n",
    "# Fig_4B:\n",
    "df_apharesis_basal_NoHighRisk_taxa_allTaxa = scan_correlation_associations(df_CART_table_taxa_basal_NoHighRisk_mean, df_mul_Aph_basal_NoHighRisk_mean, -1, apharesis_parameters, \"CART_Microbiome_Fig_4B\" )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing the abundance of top ML and differential abundance hits for CR-Abx and Survival stratifications\n",
    "* Figure 3F\n",
    "* Extended Figures 7A, 7B, 7C, 7D and 10G (taxa and pathways box-plots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bar_plots_tested(df_microbiome, abundance_threshold, df_colvec, microbiome_feature, colvec_feature,title_name, order, palette, lables, x_title, rotation = 0, showpval = True):\n",
    "    \n",
    "    df_microbiome_norm = abundance_filtration(df_microbiome, abundance_threshold)\n",
    "    \n",
    "    current_microbiome_feature_0 = df_microbiome_norm.loc[(df_colvec[colvec_feature] == 0),microbiome_feature]\n",
    "    current_microbiome_feature_1 = df_microbiome_norm.loc[(df_colvec[colvec_feature] == 1),microbiome_feature]\n",
    "    \n",
    "    statistic, pvalue = mannwhitneyu(current_microbiome_feature_0,current_microbiome_feature_1,alternative='two-sided')\n",
    "    \n",
    "    df_data = pd.concat([np.log10(df_microbiome_norm), df_colvec], axis=1)\n",
    "    \n",
    "    df_data_untransformed = pd.concat([df_microbiome_norm, df_colvec], axis=1)\n",
    "    \n",
    "    y_parameter = microbiome_feature\n",
    "    x_parameter = colvec_feature\n",
    "    \n",
    "    np.random.seed(1)\n",
    "    plt.figure(figsize = (2,5))\n",
    "    sns.set(font_scale=0.8)\n",
    "    sns.set_style(\"ticks\")\n",
    "    curr_plt = sns.boxplot(x= x_parameter, y = y_parameter, data = df_data, color = \"whitesmoke\",showfliers = False, order = order)\n",
    "    sns.stripplot(x = x_parameter, y = y_parameter, data = df_data, order = order, palette = palette)\n",
    "    plt.setp(curr_plt.get_xticklabels(), rotation=0)\n",
    "    \n",
    "    curr_plt.set_ylim([np.log10(abundance_threshold)-0.2, 0])   \n",
    "    \n",
    "    curr_plt.set_xticklabels(lables, rotation = rotation)\n",
    "    \n",
    "    if abundance_threshold == 1e-3:\n",
    "        curr_plt.set_yticks([-3, -2, -1, 0])\n",
    "        curr_plt.set_yticklabels(['1e-3', '1e-2', '1e-1', '1'])\n",
    "        \n",
    "    if abundance_threshold == 1e-4:\n",
    "        curr_plt.set_yticks([-4, -3, -2, -1, 0])\n",
    "        curr_plt.set_yticklabels(['1e-4', '1e-3', '1e-2', '1e-1', '1'])\n",
    "   \n",
    "    if showpval:\n",
    "        plt.title(microbiome_feature + \"\\n\" + \"p  = \" +str(np.round(pvalue,5)),fontsize = 12)\n",
    "    else:\n",
    "        plt.title(microbiome_feature,fontsize = 12)\n",
    "        \n",
    "    plt.xlabel(x_title,fontsize = 12)\n",
    "    plt.ylabel(\"Relative abundance\",fontsize = 12)\n",
    "    plt.savefig(path_output + title_name + \"_\" + microbiome_feature +\".pdf\",dpi=300, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    \n",
    "    if showpval == False:\n",
    "        \n",
    "        df_Pvals = pd.DataFrame()\n",
    "        \n",
    "        file_object = open(path_output + title_name+\"_\"+ microbiome_feature + \"_Pvals.txt\", 'a')\n",
    "    \n",
    "        file_object.write (title_name+\"_\"+ microbiome_feature + \"_\"+ colvec_feature)\n",
    "        file_object.write(\"\\n \")\n",
    "    \n",
    "        for paramater_a in order:\n",
    "            current_a_values = df_data_untransformed.loc[df_data_untransformed[x_parameter] == paramater_a, y_parameter]\n",
    "        \n",
    "            for paramater_b in order[order.index(paramater_a)+1:]:\n",
    "                current_b_values = df_data_untransformed.loc[df_data_untransformed[x_parameter] == paramater_b, y_parameter]\n",
    "            \n",
    "                statistic, pvalue = mannwhitneyu(current_a_values, current_b_values, alternative='two-sided')\n",
    "                \n",
    "                current_ID = paramater_a + \"------\" + paramater_b\n",
    "                df_Pvals.loc[current_ID, \"Pval\"] = pvalue  \n",
    "                \n",
    "        df_Pvals.index.name='Compared parameters'    \n",
    "                \n",
    "        padj_binary, padj =  fdrcorrection(df_Pvals.Pval, alpha=0.2, method = 'i')\n",
    "        df_Pvals.loc[:, \"Padjusted\"] = padj\n",
    "        file_object.write(\"\\n \")\n",
    "        \n",
    "        print(df_Pvals, file=file_object)\n",
    "  \n",
    "        file_object.close()\n",
    "    \n",
    "    return\n",
    "\n",
    "\n",
    "# Fig_3F:\n",
    "Taxa_to_assess_Fig_3F = [\"Bacteroides_eggerthii\", \"Ruminococcus_lactaris\", \"Eubacterium_sp_CAG_180\", \"Akkermansia_muciniphila\", \"Erysipelatoclostridium_ramosum\"]\n",
    "for current_taxon in Taxa_to_assess_Fig_3F:\n",
    "    bar_plots_tested(df_CART_table_taxa_basal_all_mean, taxa_abundance_threshold, df_colvecs_basal_all_mean, current_taxon, \"CR180_noHRAbx\", \"CART_Microbiome_Fig_3F\", \n",
    "                     order = ['noHR + CR',  \"noHR + noCR\",\"HR + CR\", \"HR + noCR\"], \n",
    "                     palette = {'noHR + CR': \"blue\", \"noHR + noCR\": \"orange\", \"HR + CR\":\"red\", \"HR + noCR\":\"green\" }, \n",
    "                     lables = ['noHR + CR',  \"noHR + noCR\",\"HR + CR\", \"HR + noCR\"], x_title= \"\", rotation = 90, showpval = False)\n",
    "\n",
    "\n",
    "# Fig_S10G:\n",
    "Taxa_to_assess_Fig_S10G = [\"Bacteroides_stercoris\", \"Bacteroides_fragilis\", \"Roseburia_faecis\", \"Eubacterium_sp_CAG_38\" ]\n",
    "for current_taxon in Taxa_to_assess_Fig_S10G:\n",
    "    bar_plots_tested(df_CART_table_taxa_basal_all_mean, taxa_abundance_threshold, df_colvecs_basal_all_mean, current_taxon, \"CR180_noHRAbx\", \"CART_Microbiome_Fig_S10G\", \n",
    "                     order = ['noHR + CR',  \"noHR + noCR\",\"HR + CR\", \"HR + noCR\"], \n",
    "                     palette = {'noHR + CR': \"blue\", \"noHR + noCR\": \"orange\", \"HR + CR\":\"red\", \"HR + noCR\":\"green\" }, \n",
    "                     lables = ['noHR + CR',  \"noHR + noCR\",\"HR + CR\", \"HR + noCR\"], x_title= \"\", rotation = 90, showpval = False)\n",
    "    \n",
    "\n",
    "# Fig_S7A (bacteria bar-plots):\n",
    "Taxa_to_assess_Fig_S7A = [\"Bifidobacterium_longum\"]\n",
    "for current_taxon in Taxa_to_assess_Fig_S7A:\n",
    "    bar_plots_tested(df_CART_table_taxa_basal_all_mean, taxa_abundance_threshold, df_colvecs_basal_all_mean, current_taxon, \"Survival_d180\", \"CART_Microbiome_Fig_S7A\", \n",
    "                     order = [1, 0], palette = {1: \"green\", 0: \"red\"}, lables = [\">180\", \"<180\"], x_title= \"Survival (days)\")\n",
    "\n",
    "    \n",
    "# Fig_S7B (bacteria bar-plots):\n",
    "Taxa_to_assess_Fig_S7B = [\"Bifidobacterium_longum\"]\n",
    "for current_taxon in Taxa_to_assess_Fig_S7B:\n",
    "    bar_plots_tested(df_CART_table_taxa_basal_NoHighRisk_mean, taxa_abundance_threshold, df_colvecs_basal_NoHighRisk_mean, current_taxon, \"Survival_d180\", \"CART_Microbiome_Fig_S7B\", \n",
    "                     order = [1, 0], palette = {1: \"green\", 0: \"red\"}, lables = [\">180\", \"<180\"], x_title= \"Survival (days)\")\n",
    "\n",
    "    \n",
    "# Fig_S7C (pathways bar-plots):\n",
    "Taxa_to_assess_Fig_S7C = [\"PWY-5265: peptidoglycan biosynthesis II (staphylococci)\"]\n",
    "for current_taxon in Taxa_to_assess_Fig_S7C:\n",
    "    bar_plots_tested(df_CART_table_pathways_basal_NoHighRisk_mean, pathways_abundance_threshold, df_colvecs_basal_NoHighRisk_mean, current_taxon, \"Survival_d180\", \"CART_Microbiome_Fig_S7C\", \n",
    "                     order = [1, 0], palette = {1: \"green\", 0: \"red\"}, lables = [\">180\", \"<180\"], x_title= \"Survival (days)\")\n",
    "\n",
    "    \n",
    "# Fig_S7D (pathways bar-plots):\n",
    "Taxa_to_assess_Fig_S7D = [\"PWY-5265: peptidoglycan biosynthesis II (staphylococci)\"]\n",
    "for current_taxon in Taxa_to_assess_Fig_S7D:\n",
    "    bar_plots_tested(df_CART_table_pathways_basal_NoHighRisk_mean, pathways_abundance_threshold, df_colvecs_basal_NoHighRisk_mean, current_taxon, \"Progression_early_binarized\", \"CART_Microbiome_Fig_S7D\", \n",
    "                     order = [0, 1], palette = {1: \"red\", 0: \"green\"}, lables = [\"NO | >180 days\", \"<180 days\"], x_title= \"\")\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing T-cells, CRP and IL-6 levels for high/low-abudance taxa stratification\n",
    "* Figures 4C, 4D and 4E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def compare_abundance_startified(df_microbiome, df_colvec, microbiome_feature, colvec_feature, title_name):\n",
    "    \n",
    "    curr_x_all = df_microbiome[microbiome_feature] \n",
    "    curr_y_all = df_colvec[colvec_feature]\n",
    "    \n",
    "    curr_x = curr_x_all[~np.isnan(curr_x_all) & ~np.isnan(curr_y_all)]\n",
    "    curr_y = curr_y_all[~np.isnan(curr_x_all) & ~np.isnan(curr_y_all)]\n",
    "    \n",
    "    startifing_threshold = np.quantile(curr_x, 0.5)\n",
    "    \n",
    "    # Stratifying based on taxa abundance and performing mannwhitneyu test:\n",
    "\n",
    "    df_data = pd.DataFrame()\n",
    "    df_data[\"x_parameter\"] = (curr_x <= startifing_threshold)\n",
    "    df_data[\"y_parameter\"] = curr_y\n",
    "    \n",
    "    colvec_feature_below = curr_y[curr_x <= startifing_threshold]\n",
    "    colvec_feature_above = curr_y[curr_x > startifing_threshold]\n",
    "    \n",
    "    statistic, pvalue = mannwhitneyu(colvec_feature_below,colvec_feature_above,alternative='two-sided')\n",
    "    \n",
    "    np.random.seed(1)\n",
    "    plt.figure(figsize = (2,5))\n",
    "    sns.set(font_scale=0.8)\n",
    "    sns.set_style(\"ticks\")\n",
    "    \n",
    "    curr_plt = sns.boxplot(x= \"x_parameter\", y = \"y_parameter\", data = df_data, color = \"whitesmoke\",showfliers = False, order = [True, False])\n",
    "    sns.stripplot(x = \"x_parameter\", y = \"y_parameter\", data = df_data, order = [True, False], palette = {1: \"green\", 0: \"red\"})\n",
    "    plt.setp(curr_plt.get_xticklabels(), rotation=0)\n",
    "  \n",
    "    plt.title(title_name + \"\\n\" + microbiome_feature + \"\\n\" + \"mannwhitneyu Pval  = \" + str(np.round(pvalue,5)),fontsize = 12)\n",
    "    curr_plt.set_xticklabels([\"Low\", \"High\"])\n",
    "    plt.xlabel(microbiome_feature)\n",
    "    plt.ylabel(colvec_feature, fontsize = 12)\n",
    "  \n",
    "    plt.savefig(path_output + title_name + \"_\" + microbiome_feature +\"_\"+colvec_feature +\"_stratified.pdf\",dpi=300, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    \n",
    "    return\n",
    "\n",
    "\n",
    "# Fig_4C:\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean, \"Lachnospira_pectinoschiza\", 'CD3_mul_Aph',  \"CART_Microbiome_Fig_4C\")\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean, \"Akkermansia_muciniphila\", 'CD4_mul_Aph',  \"CART_Microbiome_Fig_4C\")\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_NoHighRisk_mean, df_colvecs_basal_NoHighRisk_mean, \"Firmicutes_bacterium_CAG_424\", 'CD4_mul_Aph',  \"CART_Microbiome_Fig_4C\")\n",
    "\n",
    "\n",
    "# Fig_4D:\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean, \"Lachnospira_pectinoschiza\", 'CRP_depletion',  \"CART_Microbiome_Fig_4D\")\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean, \"Bacteroides_eggerthii\", 'CRP_depletion',  \"CART_Microbiome_Fig_4D\")\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean, \"Ruminococcus_lactaris\", 'CRP_depletion',  \"CART_Microbiome_Fig_4D\")\n",
    "\n",
    "\n",
    "# Fig_4E:\n",
    "compare_abundance_startified(df_CART_table_taxa_basal_all_mean, df_colvecs_basal_all_mean, \"Lachnospira_pectinoschiza\", 'IL6_depletion',  \"CART_Microbiome_Fig_4E\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alpha-diversity : Scatter plots of Shannon index versus relative sampling day\n",
    "\n",
    "* Figures 2A and 2B-(left)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def moving_median(x, y, window_size):\n",
    "    \n",
    "    x = np.array(x)\n",
    "    y = np.array(y)\n",
    "    \n",
    "    x_start = min(x)\n",
    "    x_end = max(x)\n",
    "    \n",
    "    frames = range(int(x_start + np.ceil(window_size/2)), int(x_end - np.floor(window_size/2)))\n",
    "    \n",
    "    moving_average_line = pd.DataFrame(index=frames, columns=[\"x\", \"y\"])\n",
    "    \n",
    "    for window_middle in frames:\n",
    "        \n",
    "        window_start = window_middle - window_size/2\n",
    "        window_end = window_middle + window_size/2\n",
    "        \n",
    "        current_y = y[(x > window_start) & (x < window_end )]\n",
    "        \n",
    "        if len(current_y) > 4:\n",
    "            \n",
    "            moving_average_line.loc[window_start, \"x\"] = window_middle\n",
    "            moving_average_line.loc[window_start, \"y\"] = np.median(current_y)\n",
    "       \n",
    "    return(moving_average_line)\n",
    "\n",
    "\n",
    "curr_x_all = df_CART_table_stool_all.sampleDay_sample\n",
    "curr_y_all = df_Shannon.Shannon\n",
    "\n",
    "curr_x_lim = [-20, 70]\n",
    "curr_y_lim = [curr_y_all.min()-0.1, curr_y_all.max()+0.1]\n",
    "\n",
    "curr_x = curr_x_all[(curr_x_all > curr_x_lim[0] ) & (curr_x_all < curr_x_lim[1] )]\n",
    "curr_y = curr_y_all[(curr_x_all > curr_x_lim[0] ) & (curr_x_all < curr_x_lim[1] )]\n",
    "\n",
    "curr_x_noHR = curr_x[df_CART_table_stool_all.AbxRisk_On_Sampling != \"HR\"]\n",
    "curr_y_noHR = curr_y[df_CART_table_stool_all.AbxRisk_On_Sampling != \"HR\"]\n",
    "\n",
    "curr_x_HR = curr_x[df_CART_table_stool_all.AbxRisk_On_Sampling == \"HR\"]\n",
    "curr_y_HR = curr_y[df_CART_table_stool_all.AbxRisk_On_Sampling == \"HR\"]\n",
    "\n",
    "    \n",
    "\n",
    "# Fig_2A:\n",
    "\n",
    "plt.figure(figsize = (5,5))\n",
    "\n",
    "plt.scatter(curr_x, curr_y, c= \"gray\", s = 3)\n",
    "\n",
    "moving_median_line = moving_median(curr_x, curr_y, 7)\n",
    "x1 = moving_median_line.x[~np.isnan(list(moving_median_line.x.values)) &~np.isnan(list(moving_median_line.y.values))]\n",
    "y1 = moving_median_line.y[~np.isnan(list(moving_median_line.x.values)) &~np.isnan(list(moving_median_line.y.values))]\n",
    "line_model = make_interp_spline(x1,y1, k=1)\n",
    "xs=np.linspace(min(x1),max(x1),round((max(x1) - min(x1)) / 4))\n",
    "ys=line_model(xs)\n",
    "plt.plot(xs, ys, c= \"black\")\n",
    "\n",
    "plt.xlim(curr_x_lim)\n",
    "plt.ylim(curr_y_lim)\n",
    "plt.title(\"CART_Microbiome_Fig_2A\")\n",
    "plt.xlabel(\"Days relative to CAR-T infusion\")\n",
    "plt.ylabel(\"Shannon index\")\n",
    "plt.savefig(path_output + 'CART_Microbiome_Fig_2A.pdf',dpi=300, bbox_inches='tight')\n",
    "plt.close()\n",
    "        \n",
    "\n",
    "# Fig_2B_left: \n",
    "\n",
    "plt.figure(figsize = (5,5))\n",
    "\n",
    "plt.scatter(curr_x_noHR, curr_y_noHR, c= \"green\", s = 3, label='None | LR')\n",
    "plt.scatter(curr_x_HR, curr_y_HR, c= \"red\", s = 3, label='HR')\n",
    "\n",
    "moving_median_line_noHR = moving_median(curr_x_noHR, curr_y_noHR, 7)\n",
    "x1 = moving_median_line_noHR.x[~np.isnan(list(moving_median_line_noHR.x.values)) &~np.isnan(list(moving_median_line_noHR.y.values))]\n",
    "y1 = moving_median_line_noHR.y[~np.isnan(list(moving_median_line_noHR.x.values)) &~np.isnan(list(moving_median_line_noHR.y.values))]\n",
    "line_model = make_interp_spline(x1,y1, k=1)\n",
    "xs=np.linspace(min(x1),max(x1),round((max(x1) - min(x1)) / 4))\n",
    "ys=line_model(xs)\n",
    "plt.plot(xs, ys, c= \"green\")\n",
    "\n",
    "moving_median_line_HR = moving_median(curr_x_HR, curr_y_HR, 7)\n",
    "x1 = moving_median_line_HR.x[~np.isnan(list(moving_median_line_HR.x.values)) &~np.isnan(list(moving_median_line_HR.y.values))]\n",
    "y1 = moving_median_line_HR.y[~np.isnan(list(moving_median_line_HR.x.values)) &~np.isnan(list(moving_median_line_HR.y.values))]\n",
    "line_model = make_interp_spline(x1,y1, k=1)\n",
    "xs=np.linspace(min(x1),max(x1),round((max(x1) - min(x1)) / 4))\n",
    "ys=line_model(xs)\n",
    "plt.plot(xs, ys, c= \"red\")\n",
    "\n",
    "plt.xlim(curr_x_lim)\n",
    "plt.ylim(curr_y_lim)\n",
    "plt.title(\"CART_Microbiome_Fig_2B\")\n",
    "plt.xlabel(\"Days relative to CAR-T infusion\")\n",
    "plt.ylabel(\"Shannon index\")\n",
    "plt.legend()\n",
    "plt.savefig(path_output + 'CART_Microbiome_Fig_2B_left.pdf',dpi=300, bbox_inches='tight')\n",
    "plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alpha-diversity: Box-plots comparing Shannon index between therapy outcomes\n",
    "* Figure 2C\n",
    "* Extended Figures 5B (left and right), 5C and 5D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bar_plots_Shannon(df_colvec, colvec_feature, title_name, order, palette, lables, x_title, rotation = 0, showpval = True ):\n",
    "    \n",
    "    current_microbiome_feature_0 = df_colvec.loc[(df_colvec[colvec_feature] == 0), \"Shannon\"]\n",
    "    current_microbiome_feature_1 = df_colvec.loc[(df_colvec[colvec_feature] == 1), \"Shannon\"]\n",
    "    \n",
    "    statistic, pvalue = mannwhitneyu(current_microbiome_feature_0,current_microbiome_feature_1,alternative='two-sided')\n",
    "    \n",
    "    y_parameter = \"Shannon\"\n",
    "    x_parameter = colvec_feature\n",
    "    \n",
    "    np.random.seed(1)\n",
    "    plt.figure(figsize = (2,5))\n",
    "    sns.set(font_scale=0.8)\n",
    "    sns.set_style(\"ticks\")\n",
    "    curr_plt = sns.boxplot(x= x_parameter, y = y_parameter, data = df_colvec, color = \"whitesmoke\",showfliers = False, order = order)\n",
    "    sns.stripplot(x = x_parameter, y = y_parameter, data = df_colvec, order = order, palette = palette)\n",
    "    plt.setp(curr_plt.get_xticklabels(), rotation=0)\n",
    " \n",
    "    curr_plt.set_xticklabels(lables, rotation = rotation)\n",
    "   \n",
    "    if showpval:\n",
    "        plt.title(\"Shannon\" + \"\\n\" + \"p  = \" +str(np.round(pvalue,5)),fontsize = 12)\n",
    "    else:\n",
    "        plt.title(\"Shannon\",fontsize = 12)\n",
    "        \n",
    "    plt.xlabel(x_title,fontsize = 12)\n",
    "    plt.ylabel(\"Shannon index\",fontsize = 12)\n",
    "    plt.savefig(path_output + title_name + \"_\" + \"Shannon_\"+  colvec_feature  +\".pdf\",dpi=300, bbox_inches='tight')\n",
    "    plt.close()\n",
    "    \n",
    "    if showpval == False:\n",
    "        \n",
    "        df_Pvals = pd.DataFrame()\n",
    "        \n",
    "        file_object = open(path_output + title_name+\"_\"+ \"Shannon_\" +  colvec_feature + \"_Pvals.txt\", 'a')\n",
    "    \n",
    "        file_object.write (title_name+\"_\"+ \"Shannon\" + \"_\"+ colvec_feature)\n",
    "        file_object.write(\"\\n \")\n",
    "    \n",
    "        for paramater_a in order:\n",
    "            current_a_values = df_colvec.loc[df_colvec[x_parameter] == paramater_a, y_parameter]\n",
    "        \n",
    "            for paramater_b in order[order.index(paramater_a)+1:]:\n",
    "                current_b_values = df_colvec.loc[df_colvec[x_parameter] == paramater_b, y_parameter]\n",
    "            \n",
    "                statistic, pvalue = mannwhitneyu(current_a_values, current_b_values, alternative='two-sided')\n",
    "                \n",
    "                current_ID = str(paramater_a) + \"------\" + str(paramater_b)\n",
    "                df_Pvals.loc[current_ID, \"Pval\"] = pvalue  \n",
    "                        \n",
    "        df_Pvals.index.name='Compared parameters'    \n",
    "                \n",
    "        padj_binary, padj =  fdrcorrection(df_Pvals.Pval, alpha=0.2, method = 'i')\n",
    "        df_Pvals.loc[:, \"Padjusted\"] = padj\n",
    "        file_object.write(\"\\n \")\n",
    "        \n",
    "        print(df_Pvals, file=file_object)\n",
    "        file_object.write(\"\\n \")\n",
    "        file_object.close()\n",
    "\n",
    "    return\n",
    "\n",
    "# Fig_2C (left and right):\n",
    "bar_plots_Shannon(df_colvecs_basal_all_mean, \"CR180_binarized\", \"CART_Microbiome_Fig_2C_left\",  order = [1, 0], palette = {1: \"blue\", 0: \"orange\"}, lables = [\"CR\", \"nonCR\"], x_title= \"\")\n",
    "bar_plots_Shannon(df_colvecs_basal_all_mean, \"Progression_early_binarized\", \"CART_Microbiome_Fig_2C_right\",  order = [0, 1], palette = {1: \"orange\", 0: \"blue\"}, lables = [\"No | > 180 days\", \"< 180 days\"], x_title= \"\")\n",
    "      \n",
    "    \n",
    "# Fig_S5B (left and right):\n",
    "bar_plots_Shannon(df_colvecs_basal_NoHighRisk_mean, \"CR180_binarized\", \"CART_Microbiome_Fig_S5B_left\",  order = [1, 0], palette = {1: \"blue\", 0: \"orange\"}, lables = [\"CR\", \"nonCR\"], x_title= \"Response\")\n",
    "bar_plots_Shannon(df_colvecs_basal_NoHighRisk_mean, \"Progression_early_binarized\", \"CART_Microbiome_Fig_S5B_right\",  order = [0, 1], palette = {1: \"orange\", 0: \"blue\"}, lables = [\"No | > 180 days\", \"< 180 days\"], x_title= \"Early progression\")\n",
    "\n",
    "# Fig_S5C\n",
    "bar_plots_Shannon(df_colvecs_basal_NoHighRisk_mean, \"CRS_012_34\", \"CART_Microbiome_Fig_S5C\",  order = [0, 1, 2, 34], palette = {0: \"orange\", 1: \"blue\", 2: \"red\", 34:\"gray\"}, lables = [\"0\", \"1\", \"2\", \"34\"], x_title= \"CRS\", showpval = False)\n",
    "\n",
    "# Fig_S5D:\n",
    "bar_plots_Shannon(df_colvecs_basal_NoHighRisk_mean, \"ICANS_012_34\", \"CART_Microbiome_Fig_S5D\",  order = [0, 1, 2, 34], palette = {0: \"orange\", 1: \"blue\", 2: \"red\", 34:\"gray\"}, lables = [\"0\", \"1\", \"2\", \"34\"], x_title= \"ICANS\", showpval = False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alpha-diversity: Per-patient comparison of Shannon indices before and after high-risk antibiotics\n",
    "* Figure 2B-right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stool_samples_risk_pairing = df_CART_table_stool_all.copy()\n",
    "\n",
    "for current_patient in list(set(df_CART_table_stool_all.subjectID)):\n",
    "    \n",
    "    current_patint_samples_df = df_CART_table_stool_all.loc[df_CART_table_stool_all.subjectID == current_patient]\n",
    "    current_patint_samples_df = current_patint_samples_df.sort_values(by = \"sampleDay_sample\", ascending=True)\n",
    "    \n",
    "    patient_sample_types = current_patint_samples_df.AbxRisk_On_Sampling.values\n",
    "    \n",
    "    patient_sample_types_list = list(patient_sample_types)\n",
    "    \n",
    "    if ((sum(patient_sample_types == \"HR\") > 0) & (sum(patient_sample_types != \"HR\") > 0)):\n",
    "        first_HR_sample = patient_sample_types_list.index(\"HR\")\n",
    "        \n",
    "        if (first_HR_sample > 0):\n",
    "\n",
    "            patinet_preHR_samples = list(current_patint_samples_df.stool_sample_id)[0:first_HR_sample]\n",
    "            patinet_HR_samples = list(current_patint_samples_df.stool_sample_id[current_patint_samples_df.AbxRisk_On_Sampling == \"HR\"])\n",
    "            \n",
    "            df_stool_samples_risk_pairing.loc[patinet_preHR_samples, \"LongitudinalRiskType\"] = \"noHR\"\n",
    "            df_stool_samples_risk_pairing.loc[patinet_HR_samples, \"LongitudinalRiskType\"] = \"HR\"\n",
    "        \n",
    "paired_noHR_samples = df_stool_samples_risk_pairing.LongitudinalRiskType == \"noHR\"\n",
    "paired_HR_samples = df_stool_samples_risk_pairing.LongitudinalRiskType == \"HR\"\n",
    "\n",
    "df_taxa_risk_paired_mean_noHR,  df_stool_risk_paired_mean_noHR = derive_mean_microbiome(df_CART_table_taxa_all_norm.loc[paired_noHR_samples] , df_CART_table_stool_all.loc[paired_noHR_samples])\n",
    "df_taxa_risk_paired_mean_HR, df_stool_risk_paired_mean_HR =     derive_mean_microbiome(df_CART_table_taxa_all_norm.loc[paired_HR_samples] , df_CART_table_stool_all.loc[paired_HR_samples])\n",
    "\n",
    "df_Shannon_risk_paired = pd.DataFrame(index=df_taxa_risk_paired_mean_noHR.index, columns=[\"noHR\", \"HR\"])\n",
    "\n",
    "for current_subject in df_taxa_risk_paired_mean_noHR.index:\n",
    "    df_Shannon_risk_paired.loc[current_subject, \"noHR\"] = shannon(df_taxa_risk_paired_mean_noHR.loc[current_subject,:])\n",
    "    df_Shannon_risk_paired.loc[current_subject, \"HR\"] = shannon(df_taxa_risk_paired_mean_HR.loc[current_subject,:])\n",
    "    \n",
    "statistic_sh, pvalue_sh = wilcoxon(df_Shannon_risk_paired.noHR, df_Shannon_risk_paired.HR)\n",
    "\n",
    "# Fig. 2B-right:\n",
    "np.random.seed(1)\n",
    "plt.figure(figsize = (2,5))\n",
    "sns.set(font_scale=0.8)\n",
    "sns.set_style(\"ticks\")\n",
    "curr_plt = sns.boxplot(data = df_Shannon_risk_paired, color = \"whitesmoke\",showfliers = False)\n",
    "sns.stripplot(data = df_Shannon_risk_paired, palette = {\"noHR\": \"black\", \"HR\": \"red\"})\n",
    "plt.setp(curr_plt.get_xticklabels(), rotation=0)\n",
    "plt.title(\"Pval = \" +  str(np.round(pvalue_sh,5)) + \" (n = \" + str(df_Shannon_risk_paired.shape[0])+ \")\")\n",
    "plt.ylabel(\"Shannon index (mean per patient and risk type)\")\n",
    "plt.savefig(path_output + \"CART_Microbiome_Fig_2B_right.pdf\",dpi=300, bbox_inches='tight')\n",
    "plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plots of the number of samples per relative sampling day\n",
    "* Extended Figure 5A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update(plt.rcParamsDefault)\n",
    "earlies_day = df_colvecs[\"day_colvec\"].min()\n",
    "latest_day = df_colvecs[\"day_colvec\"].max()\n",
    "all_centers = set(df_colvecs.Center)\n",
    "\n",
    "days_vec = list(range(earlies_day,latest_day+1))\n",
    "\n",
    "df_samples_per_day = pd.DataFrame(columns=all_centers, index = days_vec)\n",
    "    \n",
    "for current_center in all_centers:\n",
    "    for current_day in days_vec:\n",
    "        df_samples_per_day.loc[current_day, current_center] = sum((df_colvecs[\"Center\"] == current_center) & (df_colvecs[\"day_colvec\"] == current_day))\n",
    "            \n",
    "fig, (ax1,ax2) = plt.subplots(nrows=2, sharex=True, sharey=True,figsize=(8, 5), subplot_kw=dict(frameon=True))\n",
    "\n",
    "sbar_width = 0.8\n",
    "\n",
    "ax1.grid(color='gray', linestyle='--', linewidth=0.3)\n",
    "ax1.bar(x = days_vec, height = df_samples_per_day.HD, color =\"red\", width=sbar_width, label=\"Heidelberg\")\n",
    "ax1.bar(x = days_vec, height = df_samples_per_day.LMU, color =\"blue\", width=sbar_width, label=\"Munich\")\n",
    "ax1.bar(x = days_vec, height = df_samples_per_day.UKR,color =\"orange\", width=sbar_width,  label=\"Regensburg\")\n",
    "ax1.legend()\n",
    "ax1.set_ylim([0,20])\n",
    " \n",
    "    \n",
    "ax2.grid(color='gray', linestyle='--', linewidth=0.3)\n",
    "ax2.bar(x = days_vec, height = df_samples_per_day.MDACC, color =\"green\", width=sbar_width, label=\"MDACC\")\n",
    "ax2.bar(x = days_vec, height = df_samples_per_day.MOFFITT, color =\"purple\", width=sbar_width, label=\"Moffitt\")\n",
    "ax2.legend()\n",
    "ax2.set_ylim([0,20])\n",
    "  \n",
    "fig.text(-0.01, 0.5, 'Number of stool samples collected', va='center', rotation='vertical', fontsize = 12)\n",
    "    \n",
    "fig.text(0.08, 0.91, 'Germany', va='center', rotation='horizontal', fontsize = 14, bbox=dict(facecolor='lightgray', edgecolor='gray', pad= 5))\n",
    "fig.text(0.08, 0.46, 'US', va='center', rotation='horizontal', fontsize = 14, bbox=dict(facecolor='lightgray', edgecolor='gray', pad= 5))\n",
    "    \n",
    "    \n",
    "plt.xlabel(\"Days relative to CAR-T cell infusion\", fontsize = 12)\n",
    "plt.tight_layout()\n",
    "plt.savefig(path_output + \"CART_Microbiome_Fig_S5A.pdf\",dpi=300, bbox_inches='tight')\n",
    "plt.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
